{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "    <style>\n",
       "        .bk-notebook-logo {\n",
       "            display: block;\n",
       "            width: 20px;\n",
       "            height: 20px;\n",
       "            background-image: url(data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABQAAAAUCAYAAACNiR0NAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAABx0RVh0U29mdHdhcmUAQWRvYmUgRmlyZXdvcmtzIENTNui8sowAAAOkSURBVDiNjZRtaJVlGMd/1/08zzln5zjP1LWcU9N0NkN8m2CYjpgQYQXqSs0I84OLIC0hkEKoPtiH3gmKoiJDU7QpLgoLjLIQCpEsNJ1vqUOdO7ppbuec5+V+rj4ctwzd8IIbbi6u+8f1539dt3A78eXC7QizUF7gyV1fD1Yqg4JWz84yffhm0qkFqBogB9rM8tZdtwVsPUhWhGcFJngGeWrPzHm5oaMmkfEg1usvLFyc8jLRqDOMru7AyC8saQr7GG7f5fvDeH7Ej8CM66nIF+8yngt6HWaKh7k49Soy9nXurCi1o3qUbS3zWfrYeQDTB/Qj6kX6Ybhw4B+bOYoLKCC9H3Nu/leUTZ1JdRWkkn2ldcCamzrcf47KKXdAJllSlxAOkRgyHsGC/zRday5Qld9DyoM4/q/rUoy/CXh3jzOu3bHUVZeU+DEn8FInkPBFlu3+nW3Nw0mk6vCDiWg8CeJaxEwuHS3+z5RgY+YBR6V1Z1nxSOfoaPa4LASWxxdNp+VWTk7+4vzaou8v8PN+xo+KY2xsw6une2frhw05CTYOmQvsEhjhWjn0bmXPjpE1+kplmmkP3suftwTubK9Vq22qKmrBhpY4jvd5afdRA3wGjFAgcnTK2s4hY0/GPNIb0nErGMCRxWOOX64Z8RAC4oCXdklmEvcL8o0BfkNK4lUg9HTl+oPlQxdNo3Mg4Nv175e/1LDGzZen30MEjRUtmXSfiTVu1kK8W4txyV6BMKlbgk3lMwYCiusNy9fVfvvwMxv8Ynl6vxoByANLTWplvuj/nF9m2+PDtt1eiHPBr1oIfhCChQMBw6Aw0UulqTKZdfVvfG7VcfIqLG9bcldL/+pdWTLxLUy8Qq38heUIjh4XlzZxzQm19lLFlr8vdQ97rjZVOLf8nclzckbcD4wxXMidpX30sFd37Fv/GtwwhzhxGVAprjbg0gCAEeIgwCZyTV2Z1REEW8O4py0wsjeloKoMr6iCY6dP92H6Vw/oTyICIthibxjm/DfN9lVz8IqtqKYLUXfoKVMVQVVJOElGjrnnUt9T9wbgp8AyYKaGlqingHZU/uG2NTZSVqwHQTWkx9hxjkpWDaCg6Ckj5qebgBVbT3V3NNXMSiWSDdGV3hrtzla7J+duwPOToIg42ChPQOQjspnSlp1V+Gjdged7+8UN5CRAV7a5EdFNwCjEaBR27b3W890TE7g24NAP/mMDXRWrGoFPQI9ls/MWO2dWFAar/xcOIImbbpA3zgAAAABJRU5ErkJggg==);\n",
       "        }\n",
       "    </style>\n",
       "    <div>\n",
       "        <a href=\"https://bokeh.org\" target=\"_blank\" class=\"bk-notebook-logo\"></a>\n",
       "        <span id=\"f487e32f-a5ad-4ce4-b74e-33454321d8e2\">Loading BokehJS ...</span>\n",
       "    </div>\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": "(function(root) {\n  function now() {\n    return new Date();\n  }\n\n  const force = true;\n\n  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n    root._bokeh_onload_callbacks = [];\n    root._bokeh_is_loading = undefined;\n  }\n\nconst JS_MIME_TYPE = 'application/javascript';\n  const HTML_MIME_TYPE = 'text/html';\n  const EXEC_MIME_TYPE = 'application/vnd.bokehjs_exec.v0+json';\n  const CLASS_NAME = 'output_bokeh rendered_html';\n\n  /**\n   * Render data to the DOM node\n   */\n  function render(props, node) {\n    const script = document.createElement(\"script\");\n    node.appendChild(script);\n  }\n\n  /**\n   * Handle when an output is cleared or removed\n   */\n  function handleClearOutput(event, handle) {\n    function drop(id) {\n      const view = Bokeh.index.get_by_id(id)\n      if (view != null) {\n        view.model.document.clear()\n        Bokeh.index.delete(view)\n      }\n    }\n\n    const cell = handle.cell;\n\n    const id = cell.output_area._bokeh_element_id;\n    const server_id = cell.output_area._bokeh_server_id;\n\n    // Clean up Bokeh references\n    if (id != null) {\n      drop(id)\n    }\n\n    if (server_id !== undefined) {\n      // Clean up Bokeh references\n      const cmd_clean = \"from bokeh.io.state import curstate; print(curstate().uuid_to_server['\" + server_id + \"'].get_sessions()[0].document.roots[0]._id)\";\n      cell.notebook.kernel.execute(cmd_clean, {\n        iopub: {\n          output: function(msg) {\n            const id = msg.content.text.trim()\n            drop(id)\n          }\n        }\n      });\n      // Destroy server and session\n      const cmd_destroy = \"import bokeh.io.notebook as ion; ion.destroy_server('\" + server_id + \"')\";\n      cell.notebook.kernel.execute(cmd_destroy);\n    }\n  }\n\n  /**\n   * Handle when a new output is added\n   */\n  function handleAddOutput(event, handle) {\n    const output_area = handle.output_area;\n    const output = handle.output;\n\n    // limit handleAddOutput to display_data with EXEC_MIME_TYPE content only\n    if ((output.output_type != \"display_data\") || (!Object.prototype.hasOwnProperty.call(output.data, EXEC_MIME_TYPE))) {\n      return\n    }\n\n    const toinsert = output_area.element.find(\".\" + CLASS_NAME.split(' ')[0]);\n\n    if (output.metadata[EXEC_MIME_TYPE][\"id\"] !== undefined) {\n      toinsert[toinsert.length - 1].firstChild.textContent = output.data[JS_MIME_TYPE];\n      // store reference to embed id on output_area\n      output_area._bokeh_element_id = output.metadata[EXEC_MIME_TYPE][\"id\"];\n    }\n    if (output.metadata[EXEC_MIME_TYPE][\"server_id\"] !== undefined) {\n      const bk_div = document.createElement(\"div\");\n      bk_div.innerHTML = output.data[HTML_MIME_TYPE];\n      const script_attrs = bk_div.children[0].attributes;\n      for (let i = 0; i < script_attrs.length; i++) {\n        toinsert[toinsert.length - 1].firstChild.setAttribute(script_attrs[i].name, script_attrs[i].value);\n        toinsert[toinsert.length - 1].firstChild.textContent = bk_div.children[0].textContent\n      }\n      // store reference to server id on output_area\n      output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE][\"server_id\"];\n    }\n  }\n\n  function register_renderer(events, OutputArea) {\n\n    function append_mime(data, metadata, element) {\n      // create a DOM node to render to\n      const toinsert = this.create_output_subarea(\n        metadata,\n        CLASS_NAME,\n        EXEC_MIME_TYPE\n      );\n      this.keyboard_manager.register_events(toinsert);\n      // Render to node\n      const props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};\n      render(props, toinsert[toinsert.length - 1]);\n      element.append(toinsert);\n      return toinsert\n    }\n\n    /* Handle when an output is cleared or removed */\n    events.on('clear_output.CodeCell', handleClearOutput);\n    events.on('delete.Cell', handleClearOutput);\n\n    /* Handle when a new output is added */\n    events.on('output_added.OutputArea', handleAddOutput);\n\n    /**\n     * Register the mime type and append_mime function with output_area\n     */\n    OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {\n      /* Is output safe? */\n      safe: true,\n      /* Index of renderer in `output_area.display_order` */\n      index: 0\n    });\n  }\n\n  // register the mime type if in Jupyter Notebook environment and previously unregistered\n  if (root.Jupyter !== undefined) {\n    const events = require('base/js/events');\n    const OutputArea = require('notebook/js/outputarea').OutputArea;\n\n    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {\n      register_renderer(events, OutputArea);\n    }\n  }\n  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n    root._bokeh_timeout = Date.now() + 5000;\n    root._bokeh_failed_load = false;\n  }\n\n  const NB_LOAD_WARNING = {'data': {'text/html':\n     \"<div style='background-color: #fdd'>\\n\"+\n     \"<p>\\n\"+\n     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n     \"</p>\\n\"+\n     \"<ul>\\n\"+\n     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n     \"</ul>\\n\"+\n     \"<code>\\n\"+\n     \"from bokeh.resources import INLINE\\n\"+\n     \"output_notebook(resources=INLINE)\\n\"+\n     \"</code>\\n\"+\n     \"</div>\"}};\n\n  function display_loaded() {\n    const el = document.getElementById(\"f487e32f-a5ad-4ce4-b74e-33454321d8e2\");\n    if (el != null) {\n      el.textContent = \"BokehJS is loading...\";\n    }\n    if (root.Bokeh !== undefined) {\n      if (el != null) {\n        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n      }\n    } else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(display_loaded, 100)\n    }\n  }\n\n  function run_callbacks() {\n    try {\n      root._bokeh_onload_callbacks.forEach(function(callback) {\n        if (callback != null)\n          callback();\n      });\n    } finally {\n      delete root._bokeh_onload_callbacks\n    }\n    console.debug(\"Bokeh: all callbacks have finished\");\n  }\n\n  function load_libs(css_urls, js_urls, callback) {\n    if (css_urls == null) css_urls = [];\n    if (js_urls == null) js_urls = [];\n\n    root._bokeh_onload_callbacks.push(callback);\n    if (root._bokeh_is_loading > 0) {\n      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n      return null;\n    }\n    if (js_urls == null || js_urls.length === 0) {\n      run_callbacks();\n      return null;\n    }\n    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n    root._bokeh_is_loading = css_urls.length + js_urls.length;\n\n    function on_load() {\n      root._bokeh_is_loading--;\n      if (root._bokeh_is_loading === 0) {\n        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n        run_callbacks()\n      }\n    }\n\n    function on_error(url) {\n      console.error(\"failed to load \" + url);\n    }\n\n    for (let i = 0; i < css_urls.length; i++) {\n      const url = css_urls[i];\n      const element = document.createElement(\"link\");\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.rel = \"stylesheet\";\n      element.type = \"text/css\";\n      element.href = url;\n      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n      document.body.appendChild(element);\n    }\n\n    for (let i = 0; i < js_urls.length; i++) {\n      const url = js_urls[i];\n      const element = document.createElement('script');\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.async = false;\n      element.src = url;\n      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.head.appendChild(element);\n    }\n  };\n\n  function inject_raw_css(css) {\n    const element = document.createElement(\"style\");\n    element.appendChild(document.createTextNode(css));\n    document.body.appendChild(element);\n  }\n\n  const js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-3.3.4.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-3.3.4.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-3.3.4.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-3.3.4.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-3.3.4.min.js\"];\n  const css_urls = [];\n\n  const inline_js = [    function(Bokeh) {\n      Bokeh.set_log_level(\"info\");\n    },\nfunction(Bokeh) {\n    }\n  ];\n\n  function run_inline_js() {\n    if (root.Bokeh !== undefined || force === true) {\n          for (let i = 0; i < inline_js.length; i++) {\n      inline_js[i].call(root, root.Bokeh);\n    }\nif (force === true) {\n        display_loaded();\n      }} else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(run_inline_js, 100);\n    } else if (!root._bokeh_failed_load) {\n      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n      root._bokeh_failed_load = true;\n    } else if (force !== true) {\n      const cell = $(document.getElementById(\"f487e32f-a5ad-4ce4-b74e-33454321d8e2\")).parents('.cell').data().cell;\n      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n    }\n  }\n\n  if (root._bokeh_is_loading === 0) {\n    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n    run_inline_js();\n  } else {\n    load_libs(css_urls, js_urls, function() {\n      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n      run_inline_js();\n    });\n  }\n}(window));",
      "application/vnd.bokehjs_load.v0+json": ""
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/dc16/gentropy/.venv/lib/python3.11/site-packages/tqdm/auto.py:21: TqdmWarning:\n",
      "\n",
      "IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import pyspark.sql.functions as f\n",
    "import statsmodels.api as sm\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "from gentropy.common.session import Session\n",
    "from gentropy.dataset.colocalisation import Colocalisation\n",
    "from gentropy.dataset.l2g_prediction import L2GPrediction\n",
    "from gentropy.dataset.study_index import StudyIndex\n",
    "from gentropy.dataset.study_locus import StudyLocus\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting default log level to \"WARN\".\n",
      "To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).\n",
      "25/04/16 11:05:01 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable\n"
     ]
    }
   ],
   "source": [
    "session = Session(\n",
    "    extended_spark_conf={\n",
    "        'spark.executor.memory':'10g',\n",
    "        'spark.driver.memory':'10g'}\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "target = session.spark.read.parquet('/users/dc16/data/releases/25.03/target')\n",
    "l2g = L2GPrediction.from_parquet(\n",
    "    session, \"/users/dc16/data/releases/25.03/l2g_prediction\"\n",
    ").df.select(\"studyLocusId\", \"geneId\", \"score\")\n",
    "cred_sets = StudyLocus.from_parquet(\n",
    "    session, \"/users/dc16/data/releases/25.03/credible_set\"\n",
    ")\n",
    "studies = (\n",
    "    session.spark.read.parquet(\n",
    "        \"/users/dc16/data/gentropy_paper/gwas_study_index_therapeutic_areas\"\n",
    "    )\n",
    "    .select(\n",
    "        \"studyId\",\n",
    "        \"traitFromSource\",\n",
    "        \"traitFromSourceMappedIds\",\n",
    "        \"nCases\",\n",
    "        \"nControls\",\n",
    "        \"nSamples\",\n",
    "        \"Haematology\",\n",
    "        \"Metabolic\",\n",
    "        \"Congenital\",\n",
    "        \"Signs/symptoms\",\n",
    "        \"Neurology\",\n",
    "        \"Immune\",\n",
    "        \"Psychiatry\",\n",
    "        \"Dermatology\",\n",
    "        \"Ophthalmology\",\n",
    "        \"Cardiovascular\",\n",
    "        \"Oncology\",\n",
    "        \"Respiratory\",\n",
    "        \"Digestive\",\n",
    "        \"Endocrine\",\n",
    "        \"Musculoskeletal\",\n",
    "        \"Infection\",\n",
    "        \"Measurement\",\n",
    "        f.col(\"bianry_less_cases\").alias(\"binary_less_cases\"),\n",
    "        \"Other\",\n",
    "    )\n",
    "    .filter(f.col(\"Measurement\") == 0)\n",
    "    .filter(f.col(\"bianry_less_cases\") == 1)\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "l2g_signif = (\n",
    "    l2g.filter(f.col(\"score\") >= 0.5)\n",
    "    .join(\n",
    "        cred_sets.df.select(\n",
    "            \"studyLocusId\",\n",
    "            \"variantId\",\n",
    "            \"studyId\",\n",
    "            \"beta\",\n",
    "            \"pValueMantissa\",\n",
    "            \"pValueExponent\",\n",
    "        ),\n",
    "        \"studyLocusId\",\n",
    "        \"inner\",\n",
    "    )\n",
    "    .join(studies, \"studyId\", \"inner\")\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-RECORD 0---------------\n",
      " Haematology     | 394  \n",
      " Metabolic       | 5861 \n",
      " Congenital      | 5480 \n",
      " Signs/symptoms  | 265  \n",
      " Neurology       | 6677 \n",
      " Immune          | 6416 \n",
      " Psychiatry      | 4039 \n",
      " Dermatology     | 1535 \n",
      " Ophthalmology   | 2628 \n",
      " Cardiovascular  | 7205 \n",
      " Oncology        | 7176 \n",
      " Respiratory     | 3167 \n",
      " Digestive       | 3703 \n",
      " Endocrine       | 5644 \n",
      " Musculoskeletal | 5287 \n",
      " Infection       | 719  \n",
      " Other           | 2919 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "(\n",
    "    l2g_signif\n",
    "    .agg(\n",
    "        f.sum('Haematology').alias('Haematology'),\n",
    "        f.sum('Metabolic').alias('Metabolic'),\n",
    "        f.sum('Congenital').alias('Congenital'),\n",
    "        f.sum('Signs/symptoms').alias('Signs/symptoms'),\n",
    "        f.sum('Neurology').alias('Neurology'),\n",
    "        f.sum('Immune').alias('Immune'),\n",
    "        f.sum('Psychiatry').alias('Psychiatry'),\n",
    "        f.sum('Dermatology').alias('Dermatology'),\n",
    "        f.sum('Ophthalmology').alias('Ophthalmology'),\n",
    "        f.sum('Cardiovascular').alias('Cardiovascular'),\n",
    "        f.sum('Oncology').alias('Oncology'),\n",
    "        f.sum('Respiratory').alias('Respiratory'),\n",
    "        f.sum('Digestive').alias('Digestive'),\n",
    "        f.sum('Endocrine').alias('Endocrine'),\n",
    "        f.sum('Musculoskeletal').alias('Musculoskeletal'),\n",
    "        f.sum('Infection').alias('Infection'),\n",
    "        f.sum('Other').alias('Other'),\n",
    "    )\n",
    "    .show(vertical=True)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "studies_per_therapeutic_area = (\n",
    "    studies\n",
    "    .filter(f.col(\"Measurement\") == 0)\n",
    "    .filter(f.col(\"bianry_less_cases\") == 1)\n",
    "    .agg(\n",
    "        f.sum('Haematology').alias('Haematology'),\n",
    "        f.sum('Metabolic').alias('Metabolic'),\n",
    "        f.sum('Congenital').alias('Congenital'),\n",
    "        f.sum('Signs/symptoms').alias('Signs/symptoms'),\n",
    "        f.sum('Neurology').alias('Neurology'),\n",
    "        f.sum('Immune').alias('Immune'),\n",
    "        f.sum('Psychiatry').alias('Psychiatry'),\n",
    "        f.sum('Dermatology').alias('Dermatology'),\n",
    "        f.sum('Ophthalmology').alias('Ophthalmology'),\n",
    "        f.sum('Cardiovascular').alias('Cardiovascular'),\n",
    "        f.sum('Oncology').alias('Oncology'),\n",
    "        f.sum('Respiratory').alias('Respiratory'),\n",
    "        f.sum('Digestive').alias('Digestive'),\n",
    "        f.sum('Endocrine').alias('Endocrine'),\n",
    "        f.sum('Musculoskeletal').alias('Musculoskeletal'),\n",
    "        f.sum('Infection').alias('Infection'),\n",
    "        f.sum('Other').alias('Other'),\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_therapeutic_areas = (\n",
    "    l2g_signif.groupBy(\"geneId\")\n",
    "    .agg(\n",
    "        f.sum(\"Haematology\").alias(\"Haematology\"),\n",
    "        f.sum(\"Metabolic\").alias(\"Metabolic\"),\n",
    "        f.sum(\"Congenital\").alias(\"Congenital\"),\n",
    "        f.sum(\"Signs/symptoms\").alias(\"Signs/symptoms\"),\n",
    "        f.sum(\"Neurology\").alias(\"Neurology\"),\n",
    "        f.sum(\"Immune\").alias(\"Immune\"),\n",
    "        f.sum(\"Psychiatry\").alias(\"Psychiatry\"),\n",
    "        f.sum(\"Dermatology\").alias(\"Dermatology\"),\n",
    "        f.sum(\"Ophthalmology\").alias(\"Ophthalmology\"),\n",
    "        f.sum(\"Cardiovascular\").alias(\"Cardiovascular\"),\n",
    "        f.sum(\"Oncology\").alias(\"Oncology\"),\n",
    "        f.sum(\"Respiratory\").alias(\"Respiratory\"),\n",
    "        f.sum(\"Digestive\").alias(\"Digestive\"),\n",
    "        f.sum(\"Endocrine\").alias(\"Endocrine\"),\n",
    "        f.sum(\"Musculoskeletal\").alias(\"Musculoskeletal\"),\n",
    "        f.sum(\"Infection\").alias(\"Infection\"),\n",
    "        f.sum(\"Other\").alias(\"Other\"),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"total\",\n",
    "        f.col(\"Haematology\")\n",
    "        + f.col(\"Metabolic\")\n",
    "        + f.col(\"Congenital\")\n",
    "        + f.col(\"Signs/symptoms\")\n",
    "        + f.col(\"Neurology\")\n",
    "        + f.col(\"Immune\")\n",
    "        + f.col(\"Psychiatry\")\n",
    "        + f.col(\"Dermatology\")\n",
    "        + f.col(\"Ophthalmology\")\n",
    "        + f.col(\"Cardiovascular\")\n",
    "        + f.col(\"Oncology\")\n",
    "        + f.col(\"Respiratory\")\n",
    "        + f.col(\"Digestive\")\n",
    "        + f.col(\"Endocrine\")\n",
    "        + f.col(\"Musculoskeletal\")\n",
    "        + f.col(\"Infection\")\n",
    "        + f.col(\"Other\"),\n",
    "    )\n",
    "    .withColumns(\n",
    "        {\n",
    "            'haematology_proportion': f.col('Haematology') / studies_per_therapeutic_area.collect()[0]['Haematology'],\n",
    "            'metabolic_proportion': f.col('Metabolic') / studies_per_therapeutic_area.collect()[0]['Metabolic'],\n",
    "            'congenital_proportion': f.col('Congenital') / studies_per_therapeutic_area.collect()[0]['Congenital'],\n",
    "            'signs_symptoms_proportion': f.col('Signs/symptoms') / studies_per_therapeutic_area.collect()[0]['Signs/symptoms'],\n",
    "            'neurology_proportion': f.col('Neurology') / studies_per_therapeutic_area.collect()[0]['Neurology'],\n",
    "            'immune_proportion': f.col('Immune') / studies_per_therapeutic_area.collect()[0]['Immune'],\n",
    "            'psychiatry_proportion': f.col('Psychiatry') / studies_per_therapeutic_area.collect()[0]['Psychiatry'],\n",
    "            'dermatology_proportion': f.col('Dermatology') / studies_per_therapeutic_area.collect()[0]['Dermatology'],\n",
    "            'ophthalmology_proportion': f.col('Ophthalmology') / studies_per_therapeutic_area.collect()[0]['Ophthalmology'],\n",
    "            'cardiovascular_proportion': f.col('Cardiovascular') / studies_per_therapeutic_area.collect()[0]['Cardiovascular'],\n",
    "            'oncology_proportion': f.col('Oncology') / studies_per_therapeutic_area.collect()[0]['Oncology'],\n",
    "            'respiratory_proportion': f.col('Respiratory') / studies_per_therapeutic_area.collect()[0]['Respiratory'],\n",
    "            'digestive_proportion': f.col('Digestive') / studies_per_therapeutic_area.collect()[0]['Digestive'],\n",
    "            'endocrine_proportion': f.col('Endocrine') / studies_per_therapeutic_area.collect()[0]['Endocrine'],\n",
    "            'musculoskeletal_proportion': f.col('Musculoskeletal') / studies_per_therapeutic_area.collect()[0]['Musculoskeletal'],\n",
    "            'infection_proportion': f.col('Infection') / studies_per_therapeutic_area.collect()[0]['Infection'],\n",
    "            'other_proportion': f.col('Other') / studies_per_therapeutic_area.collect()[0]['Other'],\n",
    "        }\n",
    "    )\n",
    "    .join(\n",
    "        target.withColumn(\"constraint\", f.explode(\"constraint\"))\n",
    "        .select(\n",
    "            \"id\",\n",
    "            \"approvedSymbol\",\n",
    "            \"biotype\",\n",
    "            \"genomicLocation.chromosome\",\n",
    "            \"genomicLocation.start\",\n",
    "            \"genomicLocation.end\",\n",
    "            \"constraint.constraintType\",\n",
    "            \"constraint.score\",\n",
    "        )\n",
    "        .filter(f.col(\"constraintType\") == \"lof\"),\n",
    "        target[\"id\"] == f.col(\"geneId\"),\n",
    "        \"inner\",\n",
    "    )\n",
    "    .drop(\"id\", \"constraintType\")\n",
    "    .withColumnRenamed(\"score\", \"lofConstraint\")\n",
    "    .withColumn(\n",
    "        'pleiotropy',\n",
    "        (\n",
    "            f.when(f.col(\"Haematology\") > 0, 1).otherwise(0)\n",
    "            + f.when(f.col(\"Metabolic\") > 0, 1).otherwise(0)\n",
    "            + f.when(f.col(\"Congenital\") > 0, 1).otherwise(0)\n",
    "            + f.when(f.col(\"Signs/symptoms\") > 0, 1).otherwise(0)\n",
    "            + f.when(f.col(\"Neurology\") > 0, 1).otherwise(0)\n",
    "            + f.when(f.col(\"Immune\") > 0, 1).otherwise(0)\n",
    "            + f.when(f.col(\"Psychiatry\") > 0, 1).otherwise(0)\n",
    "            + f.when(f.col(\"Dermatology\") > 0, 1).otherwise(0)\n",
    "            + f.when(f.col(\"Ophthalmology\") > 0, 1).otherwise(0)\n",
    "            + f.when(f.col(\"Cardiovascular\") > 0, 1).otherwise(0)\n",
    "            + f.when(f.col(\"Oncology\") > 0, 1).otherwise(0)\n",
    "            + f.when(f.col(\"Respiratory\") > 0, 1).otherwise(0)\n",
    "            + f.when(f.col(\"Digestive\") > 0, 1).otherwise(0)\n",
    "            + f.when(f.col(\"Endocrine\") > 0, 1).otherwise(0)\n",
    "            + f.when(f.col(\"Musculoskeletal\") > 0, 1).otherwise(0)\n",
    "            + f.when(f.col(\"Infection\") > 0, 1).otherwise(0)\n",
    "            + f.when(f.col(\"Other\") > 0, 1).otherwise(0)\n",
    "        )\n",
    "        / 17\n",
    "    )\n",
    "    .sort(f.desc(\"total\"))\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "25/04/16 11:05:11 WARN MetricsConfig: Cannot locate configuration: tried hadoop-metrics2-google-hadoop-file-system.properties,hadoop-metrics2.properties\n",
      "25/04/16 11:05:11 WARN SparkStringUtils: Truncated the string representation of a plan since it was too large. This behavior can be adjusted by setting 'spark.sql.debug.maxToStringFields'.\n",
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "genes_therapeutic_areas.write.parquet('gs://genetics-portal-dev-analysis/dc16/output/pleiotropy_genes_therapeutic_areas', mode='overwrite')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of protein-coding genes with a loss-of-function constraint score assigned to a therapeutic area: 6834\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+------------------+\n",
      "|avg(lofConstraint)|\n",
      "+------------------+\n",
      "| 0.302645954727438|\n",
      "+------------------+\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print('Number of protein-coding genes with a loss-of-function constraint score assigned to a therapeutic area:', genes_therapeutic_areas.count())\n",
    "genes_therapeutic_areas.agg(f.mean('lofConstraint')).show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of protein-coding genes with a loss-of-function constraint score not assigned to a therapeutic area 11765\n",
      "+-------------------+\n",
      "|         avg(score)|\n",
      "+-------------------+\n",
      "|0.21912208931020252|\n",
      "+-------------------+\n",
      "\n"
     ]
    }
   ],
   "source": [
    "genes_not_assigned =(\n",
    "    target\n",
    "    .filter(~f.col('id').isin([row.geneId for row in genes_therapeutic_areas.select('geneId').collect()]))\n",
    "    .filter(f.col('biotype') == 'protein_coding')\n",
    "    .withColumn('constraint', f.explode('constraint'))\n",
    "    .select(\n",
    "        'id',\n",
    "        'approvedSymbol',\n",
    "        'constraint.constraintType',\n",
    "        'constraint.score',\n",
    "    )\n",
    "    .filter(f.col('constraintType') == 'lof')\n",
    "    .drop('constraintType')\n",
    ")\n",
    "print(\"Number of protein-coding genes with a loss-of-function constraint score not assigned to a therapeutic area:\", genes_not_assigned.count())\n",
    "genes_not_assigned.agg(f.mean('score')).show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "drugs = (\n",
    "    session.spark.read.parquet('gs://open-targets-data-releases/25.03/output/drug_molecule')\n",
    "    .select('id', 'blackBoxWarning', 'maximumClinicalTrialPhase', 'hasBeenWithdrawn', f.explode('linkedTargets.rows').alias('linkedTargets'))\n",
    ")\n",
    "genes_drugs = (\n",
    "    genes_therapeutic_areas\n",
    "    .join(\n",
    "        drugs,\n",
    "        genes_therapeutic_areas['geneId'] == drugs['linkedTargets'],\n",
    "        'inner'\n",
    "    )\n",
    "    .drop('linkedTargets')\n",
    "    .filter(f.col('maximumClinicalTrialPhase').isin([3.0, 4.0]))\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "genes_pd = genes_therapeutic_areas.filter(f.col('lofConstraint').isNotNull()).toPandas()\n",
    "genes_drugs_pd = genes_drugs.filter(f.col('lofConstraint').isNotNull()).toPandas()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneId</th>\n",
       "      <th>Haematology</th>\n",
       "      <th>Metabolic</th>\n",
       "      <th>Congenital</th>\n",
       "      <th>Signs/symptoms</th>\n",
       "      <th>Neurology</th>\n",
       "      <th>Immune</th>\n",
       "      <th>Psychiatry</th>\n",
       "      <th>Dermatology</th>\n",
       "      <th>Ophthalmology</th>\n",
       "      <th>...</th>\n",
       "      <th>musculoskeletal_proportion</th>\n",
       "      <th>infection_proportion</th>\n",
       "      <th>other_proportion</th>\n",
       "      <th>approvedSymbol</th>\n",
       "      <th>biotype</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>lofConstraint</th>\n",
       "      <th>pleiotropy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000147883</td>\n",
       "      <td>0</td>\n",
       "      <td>73</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>38</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>49</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.006369</td>\n",
       "      <td>CDKN2B</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>9</td>\n",
       "      <td>22002903</td>\n",
       "      <td>22009305</td>\n",
       "      <td>7.336900e-03</td>\n",
       "      <td>0.764706</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000130203</td>\n",
       "      <td>0</td>\n",
       "      <td>33</td>\n",
       "      <td>29</td>\n",
       "      <td>4</td>\n",
       "      <td>77</td>\n",
       "      <td>1</td>\n",
       "      <td>60</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>...</td>\n",
       "      <td>0.005056</td>\n",
       "      <td>0.011994</td>\n",
       "      <td>0.003017</td>\n",
       "      <td>APOE</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>19</td>\n",
       "      <td>44905791</td>\n",
       "      <td>44909393</td>\n",
       "      <td>1.868500e-03</td>\n",
       "      <td>0.941176</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000164093</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>26</td>\n",
       "      <td>0</td>\n",
       "      <td>38</td>\n",
       "      <td>24</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>...</td>\n",
       "      <td>0.003933</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002682</td>\n",
       "      <td>PITX2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>4</td>\n",
       "      <td>110617423</td>\n",
       "      <td>110642123</td>\n",
       "      <td>9.761800e-01</td>\n",
       "      <td>0.647059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000134242</td>\n",
       "      <td>3</td>\n",
       "      <td>32</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>93</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>...</td>\n",
       "      <td>0.028652</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001006</td>\n",
       "      <td>PTPN22</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1</td>\n",
       "      <td>113813811</td>\n",
       "      <td>113871753</td>\n",
       "      <td>3.492100e-21</td>\n",
       "      <td>0.705882</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000162594</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>82</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>92</td>\n",
       "      <td>0</td>\n",
       "      <td>13</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>0.006180</td>\n",
       "      <td>0.002999</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>IL23R</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1</td>\n",
       "      <td>67138907</td>\n",
       "      <td>67259979</td>\n",
       "      <td>5.596100e-04</td>\n",
       "      <td>0.647059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6794</th>\n",
       "      <td>ENSG00000111145</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>ELK3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>12</td>\n",
       "      <td>96194375</td>\n",
       "      <td>96269824</td>\n",
       "      <td>8.350800e-02</td>\n",
       "      <td>0.058824</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6795</th>\n",
       "      <td>ENSG00000141905</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NFIC</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>19</td>\n",
       "      <td>3314403</td>\n",
       "      <td>3469217</td>\n",
       "      <td>9.053400e-02</td>\n",
       "      <td>0.058824</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6796</th>\n",
       "      <td>ENSG00000177189</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>RPS6KA3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>X</td>\n",
       "      <td>20149911</td>\n",
       "      <td>20267519</td>\n",
       "      <td>9.999900e-01</td>\n",
       "      <td>0.058824</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6797</th>\n",
       "      <td>ENSG00000123094</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>RASSF8</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>12</td>\n",
       "      <td>25958232</td>\n",
       "      <td>26079892</td>\n",
       "      <td>9.415800e-01</td>\n",
       "      <td>0.058824</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6798</th>\n",
       "      <td>ENSG00000025708</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>TYMP</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>22</td>\n",
       "      <td>50525752</td>\n",
       "      <td>50530032</td>\n",
       "      <td>1.169400e-04</td>\n",
       "      <td>0.058824</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6799 rows Ã— 43 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               geneId  Haematology  Metabolic  Congenital  Signs/symptoms  \\\n",
       "0     ENSG00000147883            0         73           5               0   \n",
       "1     ENSG00000130203            0         33          29               4   \n",
       "2     ENSG00000164093            0          0          26               0   \n",
       "3     ENSG00000134242            3         32          13               0   \n",
       "4     ENSG00000162594            0          2          82               0   \n",
       "...               ...          ...        ...         ...             ...   \n",
       "6794  ENSG00000111145            0          0           0               0   \n",
       "6795  ENSG00000141905            0          0           0               0   \n",
       "6796  ENSG00000177189            0          0           0               0   \n",
       "6797  ENSG00000123094            0          0           0               0   \n",
       "6798  ENSG00000025708            0          0           0               0   \n",
       "\n",
       "      Neurology  Immune  Psychiatry  Dermatology  Ophthalmology  ...  \\\n",
       "0            38       2           4            1             49  ...   \n",
       "1            77       1          60            1             12  ...   \n",
       "2            38      24           2            0              9  ...   \n",
       "3            11      93           0            0              7  ...   \n",
       "4             2      92           0           13              3  ...   \n",
       "...         ...     ...         ...          ...            ...  ...   \n",
       "6794          0       0           0            0              0  ...   \n",
       "6795          0       0           0            0              0  ...   \n",
       "6796          0       0           0            0              0  ...   \n",
       "6797          0       0           0            0              0  ...   \n",
       "6798          0       0           0            0              0  ...   \n",
       "\n",
       "      musculoskeletal_proportion  infection_proportion  other_proportion  \\\n",
       "0                       0.000000              0.000000          0.006369   \n",
       "1                       0.005056              0.011994          0.003017   \n",
       "2                       0.003933              0.000000          0.002682   \n",
       "3                       0.028652              0.000000          0.001006   \n",
       "4                       0.006180              0.002999          0.000000   \n",
       "...                          ...                   ...               ...   \n",
       "6794                    0.000000              0.000000          0.000000   \n",
       "6795                    0.000000              0.000000          0.000000   \n",
       "6796                    0.000000              0.000000          0.000000   \n",
       "6797                    0.000000              0.000000          0.000000   \n",
       "6798                    0.000000              0.000000          0.000000   \n",
       "\n",
       "      approvedSymbol         biotype  chromosome      start        end  \\\n",
       "0             CDKN2B  protein_coding           9   22002903   22009305   \n",
       "1               APOE  protein_coding          19   44905791   44909393   \n",
       "2              PITX2  protein_coding           4  110617423  110642123   \n",
       "3             PTPN22  protein_coding           1  113813811  113871753   \n",
       "4              IL23R  protein_coding           1   67138907   67259979   \n",
       "...              ...             ...         ...        ...        ...   \n",
       "6794            ELK3  protein_coding          12   96194375   96269824   \n",
       "6795            NFIC  protein_coding          19    3314403    3469217   \n",
       "6796         RPS6KA3  protein_coding           X   20149911   20267519   \n",
       "6797          RASSF8  protein_coding          12   25958232   26079892   \n",
       "6798            TYMP  protein_coding          22   50525752   50530032   \n",
       "\n",
       "      lofConstraint  pleiotropy  \n",
       "0      7.336900e-03    0.764706  \n",
       "1      1.868500e-03    0.941176  \n",
       "2      9.761800e-01    0.647059  \n",
       "3      3.492100e-21    0.705882  \n",
       "4      5.596100e-04    0.647059  \n",
       "...             ...         ...  \n",
       "6794   8.350800e-02    0.058824  \n",
       "6795   9.053400e-02    0.058824  \n",
       "6796   9.999900e-01    0.058824  \n",
       "6797   9.415800e-01    0.058824  \n",
       "6798   1.169400e-04    0.058824  \n",
       "\n",
       "[6799 rows x 43 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                            OLS Regression Results                            \n",
      "==============================================================================\n",
      "Dep. Variable:          lofConstraint   R-squared:                       0.022\n",
      "Model:                            OLS   Adj. R-squared:                  0.021\n",
      "Method:                 Least Squares   F-statistic:                     150.1\n",
      "Date:                Wed, 16 Apr 2025   Prob (F-statistic):           3.80e-34\n",
      "Time:                        11:05:32   Log-Likelihood:                -3602.7\n",
      "No. Observations:                6799   AIC:                             7209.\n",
      "Df Residuals:                    6797   BIC:                             7223.\n",
      "Df Model:                           1                                         \n",
      "Covariance Type:            nonrobust                                         \n",
      "==============================================================================\n",
      "                 coef    std err          t      P>|t|      [0.025      0.975]\n",
      "------------------------------------------------------------------------------\n",
      "const          0.2220      0.008     26.877      0.000       0.206       0.238\n",
      "pleiotropy     0.4266      0.035     12.251      0.000       0.358       0.495\n",
      "==============================================================================\n",
      "Omnibus:                     2586.356   Durbin-Watson:                   2.024\n",
      "Prob(Omnibus):                  0.000   Jarque-Bera (JB):             1090.535\n",
      "Skew:                           0.830   Prob(JB):                    1.56e-237\n",
      "Kurtosis:                       1.955   Cond. No.                         7.24\n",
      "==============================================================================\n",
      "\n",
      "Notes:\n",
      "[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.\n"
     ]
    }
   ],
   "source": [
    "x = genes_pd['pleiotropy']\n",
    "x = sm.add_constant(x)\n",
    "y = genes_pd['lofConstraint']\n",
    "print(sm.OLS(y, x).fit().summary())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                            OLS Regression Results                            \n",
      "==============================================================================\n",
      "Dep. Variable:          lofConstraint   R-squared:                       0.015\n",
      "Model:                            OLS   Adj. R-squared:                  0.012\n",
      "Method:                 Least Squares   F-statistic:                     5.994\n",
      "Date:                Wed, 16 Apr 2025   Prob (F-statistic):           5.20e-14\n",
      "Time:                        11:05:32   Log-Likelihood:                -3626.3\n",
      "No. Observations:                6799   AIC:                             7289.\n",
      "Df Residuals:                    6781   BIC:                             7411.\n",
      "Df Model:                          17                                         \n",
      "Covariance Type:            nonrobust                                         \n",
      "==============================================================================================\n",
      "                                 coef    std err          t      P>|t|      [0.025      0.975]\n",
      "----------------------------------------------------------------------------------------------\n",
      "const                          0.2803      0.006     48.832      0.000       0.269       0.292\n",
      "haematology_proportion         2.0800      3.305      0.629      0.529      -4.400       8.560\n",
      "metabolic_proportion          -8.4268      2.143     -3.933      0.000     -12.627      -4.227\n",
      "congenital_proportion          1.9051      4.577      0.416      0.677      -7.068      10.878\n",
      "signs_symptoms_proportion      8.7639      7.055      1.242      0.214      -5.067      22.595\n",
      "neurology_proportion          13.3209     10.563      1.261      0.207      -7.386      34.028\n",
      "immune_proportion             -6.5511      3.513     -1.865      0.062     -13.438       0.335\n",
      "psychiatry_proportion          9.5514      5.770      1.655      0.098      -1.760      20.862\n",
      "dermatology_proportion         6.7534      4.232      1.596      0.111      -1.543      15.050\n",
      "ophthalmology_proportion       1.4780      2.243      0.659      0.510      -2.918       5.874\n",
      "cardiovascular_proportion      3.8758      2.107      1.839      0.066      -0.255       8.007\n",
      "oncology_proportion            2.5524      2.275      1.122      0.262      -1.908       7.012\n",
      "respiratory_proportion         0.7949      1.593      0.499      0.618      -2.328       3.918\n",
      "digestive_proportion           1.7967      5.219      0.344      0.731      -8.434      12.028\n",
      "endocrine_proportion          11.3261      2.596      4.362      0.000       6.236      16.416\n",
      "musculoskeletal_proportion    14.3917      5.022      2.866      0.004       4.547      24.236\n",
      "infection_proportion          -8.0012      3.491     -2.292      0.022     -14.844      -1.158\n",
      "other_proportion              13.5210     11.065      1.222      0.222      -8.170      35.212\n",
      "==============================================================================\n",
      "Omnibus:                     2723.686   Durbin-Watson:                   2.018\n",
      "Prob(Omnibus):                  0.000   Jarque-Bera (JB):             1099.345\n",
      "Skew:                           0.831   Prob(JB):                    1.91e-239\n",
      "Kurtosis:                       1.944   Cond. No.                     2.36e+03\n",
      "==============================================================================\n",
      "\n",
      "Notes:\n",
      "[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.\n",
      "[2] The condition number is large, 2.36e+03. This might indicate that there are\n",
      "strong multicollinearity or other numerical problems.\n"
     ]
    }
   ],
   "source": [
    "x = genes_pd[['haematology_proportion', 'metabolic_proportion',\n",
    "       'congenital_proportion', 'signs_symptoms_proportion',\n",
    "       'neurology_proportion', 'immune_proportion', 'psychiatry_proportion',\n",
    "       'dermatology_proportion', 'ophthalmology_proportion',\n",
    "       'cardiovascular_proportion', 'oncology_proportion',\n",
    "       'respiratory_proportion', 'digestive_proportion',\n",
    "       'endocrine_proportion', 'musculoskeletal_proportion',\n",
    "       'infection_proportion', 'other_proportion']]\n",
    "x = sm.add_constant(x)\n",
    "y = genes_pd['lofConstraint']\n",
    "print(sm.OLS(y, x).fit().summary())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Optimization terminated successfully.\n",
      "         Current function value: 0.544019\n",
      "         Iterations 5\n",
      "                           Logit Regression Results                           \n",
      "==============================================================================\n",
      "Dep. Variable:        blackBoxWarning   No. Observations:                 5588\n",
      "Model:                          Logit   Df Residuals:                     5585\n",
      "Method:                           MLE   Df Model:                            2\n",
      "Date:                Wed, 16 Apr 2025   Pseudo R-squ.:                0.004109\n",
      "Time:                        11:05:33   Log-Likelihood:                -3040.0\n",
      "converged:                       True   LL-Null:                       -3052.5\n",
      "Covariance Type:            nonrobust   LLR p-value:                 3.576e-06\n",
      "=================================================================================\n",
      "                    coef    std err          z      P>|z|      [0.025      0.975]\n",
      "---------------------------------------------------------------------------------\n",
      "const            -1.0710      0.059    -18.114      0.000      -1.187      -0.955\n",
      "pleiotropy       -1.0159      0.228     -4.456      0.000      -1.463      -0.569\n",
      "lofConstraint     0.1937      0.072      2.694      0.007       0.053       0.335\n",
      "=================================================================================\n"
     ]
    }
   ],
   "source": [
    "x = genes_drugs_pd[['pleiotropy', 'lofConstraint']]\n",
    "x = sm.add_constant(x)\n",
    "y = genes_drugs_pd['blackBoxWarning'].astype(int)\n",
    "print(sm.Logit(y, x).fit().summary())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Optimization terminated successfully.\n",
      "         Current function value: 0.239517\n",
      "         Iterations 7\n",
      "                           Logit Regression Results                           \n",
      "==============================================================================\n",
      "Dep. Variable:       hasBeenWithdrawn   No. Observations:                 5588\n",
      "Model:                          Logit   Df Residuals:                     5585\n",
      "Method:                           MLE   Df Model:                            2\n",
      "Date:                Wed, 16 Apr 2025   Pseudo R-squ.:                 0.01346\n",
      "Time:                        11:05:33   Log-Likelihood:                -1338.4\n",
      "converged:                       True   LL-Null:                       -1356.7\n",
      "Covariance Type:            nonrobust   LLR p-value:                 1.181e-08\n",
      "=================================================================================\n",
      "                    coef    std err          z      P>|z|      [0.025      0.975]\n",
      "---------------------------------------------------------------------------------\n",
      "const            -2.3504      0.104    -22.517      0.000      -2.555      -2.146\n",
      "pleiotropy       -2.4986      0.468     -5.339      0.000      -3.416      -1.581\n",
      "lofConstraint     0.2961      0.123      2.409      0.016       0.055       0.537\n",
      "=================================================================================\n"
     ]
    }
   ],
   "source": [
    "x = genes_drugs_pd[['pleiotropy', 'lofConstraint']]\n",
    "x = sm.add_constant(x)\n",
    "y = genes_drugs_pd['hasBeenWithdrawn'].astype(int)\n",
    "print(sm.Logit(y, x).fit().summary())"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
