{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "    <style>\n",
       "        .bk-notebook-logo {\n",
       "            display: block;\n",
       "            width: 20px;\n",
       "            height: 20px;\n",
       "            background-image: url(data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABQAAAAUCAYAAACNiR0NAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAABx0RVh0U29mdHdhcmUAQWRvYmUgRmlyZXdvcmtzIENTNui8sowAAAOkSURBVDiNjZRtaJVlGMd/1/08zzln5zjP1LWcU9N0NkN8m2CYjpgQYQXqSs0I84OLIC0hkEKoPtiH3gmKoiJDU7QpLgoLjLIQCpEsNJ1vqUOdO7ppbuec5+V+rj4ctwzd8IIbbi6u+8f1539dt3A78eXC7QizUF7gyV1fD1Yqg4JWz84yffhm0qkFqBogB9rM8tZdtwVsPUhWhGcFJngGeWrPzHm5oaMmkfEg1usvLFyc8jLRqDOMru7AyC8saQr7GG7f5fvDeH7Ej8CM66nIF+8yngt6HWaKh7k49Soy9nXurCi1o3qUbS3zWfrYeQDTB/Qj6kX6Ybhw4B+bOYoLKCC9H3Nu/leUTZ1JdRWkkn2ldcCamzrcf47KKXdAJllSlxAOkRgyHsGC/zRday5Qld9DyoM4/q/rUoy/CXh3jzOu3bHUVZeU+DEn8FInkPBFlu3+nW3Nw0mk6vCDiWg8CeJaxEwuHS3+z5RgY+YBR6V1Z1nxSOfoaPa4LASWxxdNp+VWTk7+4vzaou8v8PN+xo+KY2xsw6une2frhw05CTYOmQvsEhjhWjn0bmXPjpE1+kplmmkP3suftwTubK9Vq22qKmrBhpY4jvd5afdRA3wGjFAgcnTK2s4hY0/GPNIb0nErGMCRxWOOX64Z8RAC4oCXdklmEvcL8o0BfkNK4lUg9HTl+oPlQxdNo3Mg4Nv175e/1LDGzZen30MEjRUtmXSfiTVu1kK8W4txyV6BMKlbgk3lMwYCiusNy9fVfvvwMxv8Ynl6vxoByANLTWplvuj/nF9m2+PDtt1eiHPBr1oIfhCChQMBw6Aw0UulqTKZdfVvfG7VcfIqLG9bcldL/+pdWTLxLUy8Qq38heUIjh4XlzZxzQm19lLFlr8vdQ97rjZVOLf8nclzckbcD4wxXMidpX30sFd37Fv/GtwwhzhxGVAprjbg0gCAEeIgwCZyTV2Z1REEW8O4py0wsjeloKoMr6iCY6dP92H6Vw/oTyICIthibxjm/DfN9lVz8IqtqKYLUXfoKVMVQVVJOElGjrnnUt9T9wbgp8AyYKaGlqingHZU/uG2NTZSVqwHQTWkx9hxjkpWDaCg6Ckj5qebgBVbT3V3NNXMSiWSDdGV3hrtzla7J+duwPOToIg42ChPQOQjspnSlp1V+Gjdged7+8UN5CRAV7a5EdFNwCjEaBR27b3W890TE7g24NAP/mMDXRWrGoFPQI9ls/MWO2dWFAar/xcOIImbbpA3zgAAAABJRU5ErkJggg==);\n",
       "        }\n",
       "    </style>\n",
       "    <div>\n",
       "        <a href=\"https://bokeh.org\" target=\"_blank\" class=\"bk-notebook-logo\"></a>\n",
       "        <span id=\"ea9b634a-6886-431d-a6f9-3cd0086eb6da\">Loading BokehJS ...</span>\n",
       "    </div>\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": "(function(root) {\n  function now() {\n    return new Date();\n  }\n\n  const force = true;\n\n  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n    root._bokeh_onload_callbacks = [];\n    root._bokeh_is_loading = undefined;\n  }\n\nconst JS_MIME_TYPE = 'application/javascript';\n  const HTML_MIME_TYPE = 'text/html';\n  const EXEC_MIME_TYPE = 'application/vnd.bokehjs_exec.v0+json';\n  const CLASS_NAME = 'output_bokeh rendered_html';\n\n  /**\n   * Render data to the DOM node\n   */\n  function render(props, node) {\n    const script = document.createElement(\"script\");\n    node.appendChild(script);\n  }\n\n  /**\n   * Handle when an output is cleared or removed\n   */\n  function handleClearOutput(event, handle) {\n    function drop(id) {\n      const view = Bokeh.index.get_by_id(id)\n      if (view != null) {\n        view.model.document.clear()\n        Bokeh.index.delete(view)\n      }\n    }\n\n    const cell = handle.cell;\n\n    const id = cell.output_area._bokeh_element_id;\n    const server_id = cell.output_area._bokeh_server_id;\n\n    // Clean up Bokeh references\n    if (id != null) {\n      drop(id)\n    }\n\n    if (server_id !== undefined) {\n      // Clean up Bokeh references\n      const cmd_clean = \"from bokeh.io.state import curstate; print(curstate().uuid_to_server['\" + server_id + \"'].get_sessions()[0].document.roots[0]._id)\";\n      cell.notebook.kernel.execute(cmd_clean, {\n        iopub: {\n          output: function(msg) {\n            const id = msg.content.text.trim()\n            drop(id)\n          }\n        }\n      });\n      // Destroy server and session\n      const cmd_destroy = \"import bokeh.io.notebook as ion; ion.destroy_server('\" + server_id + \"')\";\n      cell.notebook.kernel.execute(cmd_destroy);\n    }\n  }\n\n  /**\n   * Handle when a new output is added\n   */\n  function handleAddOutput(event, handle) {\n    const output_area = handle.output_area;\n    const output = handle.output;\n\n    // limit handleAddOutput to display_data with EXEC_MIME_TYPE content only\n    if ((output.output_type != \"display_data\") || (!Object.prototype.hasOwnProperty.call(output.data, EXEC_MIME_TYPE))) {\n      return\n    }\n\n    const toinsert = output_area.element.find(\".\" + CLASS_NAME.split(' ')[0]);\n\n    if (output.metadata[EXEC_MIME_TYPE][\"id\"] !== undefined) {\n      toinsert[toinsert.length - 1].firstChild.textContent = output.data[JS_MIME_TYPE];\n      // store reference to embed id on output_area\n      output_area._bokeh_element_id = output.metadata[EXEC_MIME_TYPE][\"id\"];\n    }\n    if (output.metadata[EXEC_MIME_TYPE][\"server_id\"] !== undefined) {\n      const bk_div = document.createElement(\"div\");\n      bk_div.innerHTML = output.data[HTML_MIME_TYPE];\n      const script_attrs = bk_div.children[0].attributes;\n      for (let i = 0; i < script_attrs.length; i++) {\n        toinsert[toinsert.length - 1].firstChild.setAttribute(script_attrs[i].name, script_attrs[i].value);\n        toinsert[toinsert.length - 1].firstChild.textContent = bk_div.children[0].textContent\n      }\n      // store reference to server id on output_area\n      output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE][\"server_id\"];\n    }\n  }\n\n  function register_renderer(events, OutputArea) {\n\n    function append_mime(data, metadata, element) {\n      // create a DOM node to render to\n      const toinsert = this.create_output_subarea(\n        metadata,\n        CLASS_NAME,\n        EXEC_MIME_TYPE\n      );\n      this.keyboard_manager.register_events(toinsert);\n      // Render to node\n      const props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};\n      render(props, toinsert[toinsert.length - 1]);\n      element.append(toinsert);\n      return toinsert\n    }\n\n    /* Handle when an output is cleared or removed */\n    events.on('clear_output.CodeCell', handleClearOutput);\n    events.on('delete.Cell', handleClearOutput);\n\n    /* Handle when a new output is added */\n    events.on('output_added.OutputArea', handleAddOutput);\n\n    /**\n     * Register the mime type and append_mime function with output_area\n     */\n    OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {\n      /* Is output safe? */\n      safe: true,\n      /* Index of renderer in `output_area.display_order` */\n      index: 0\n    });\n  }\n\n  // register the mime type if in Jupyter Notebook environment and previously unregistered\n  if (root.Jupyter !== undefined) {\n    const events = require('base/js/events');\n    const OutputArea = require('notebook/js/outputarea').OutputArea;\n\n    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {\n      register_renderer(events, OutputArea);\n    }\n  }\n  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n    root._bokeh_timeout = Date.now() + 5000;\n    root._bokeh_failed_load = false;\n  }\n\n  const NB_LOAD_WARNING = {'data': {'text/html':\n     \"<div style='background-color: #fdd'>\\n\"+\n     \"<p>\\n\"+\n     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n     \"</p>\\n\"+\n     \"<ul>\\n\"+\n     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n     \"</ul>\\n\"+\n     \"<code>\\n\"+\n     \"from bokeh.resources import INLINE\\n\"+\n     \"output_notebook(resources=INLINE)\\n\"+\n     \"</code>\\n\"+\n     \"</div>\"}};\n\n  function display_loaded() {\n    const el = document.getElementById(\"ea9b634a-6886-431d-a6f9-3cd0086eb6da\");\n    if (el != null) {\n      el.textContent = \"BokehJS is loading...\";\n    }\n    if (root.Bokeh !== undefined) {\n      if (el != null) {\n        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n      }\n    } else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(display_loaded, 100)\n    }\n  }\n\n  function run_callbacks() {\n    try {\n      root._bokeh_onload_callbacks.forEach(function(callback) {\n        if (callback != null)\n          callback();\n      });\n    } finally {\n      delete root._bokeh_onload_callbacks\n    }\n    console.debug(\"Bokeh: all callbacks have finished\");\n  }\n\n  function load_libs(css_urls, js_urls, callback) {\n    if (css_urls == null) css_urls = [];\n    if (js_urls == null) js_urls = [];\n\n    root._bokeh_onload_callbacks.push(callback);\n    if (root._bokeh_is_loading > 0) {\n      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n      return null;\n    }\n    if (js_urls == null || js_urls.length === 0) {\n      run_callbacks();\n      return null;\n    }\n    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n    root._bokeh_is_loading = css_urls.length + js_urls.length;\n\n    function on_load() {\n      root._bokeh_is_loading--;\n      if (root._bokeh_is_loading === 0) {\n        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n        run_callbacks()\n      }\n    }\n\n    function on_error(url) {\n      console.error(\"failed to load \" + url);\n    }\n\n    for (let i = 0; i < css_urls.length; i++) {\n      const url = css_urls[i];\n      const element = document.createElement(\"link\");\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.rel = \"stylesheet\";\n      element.type = \"text/css\";\n      element.href = url;\n      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n      document.body.appendChild(element);\n    }\n\n    for (let i = 0; i < js_urls.length; i++) {\n      const url = js_urls[i];\n      const element = document.createElement('script');\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.async = false;\n      element.src = url;\n      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.head.appendChild(element);\n    }\n  };\n\n  function inject_raw_css(css) {\n    const element = document.createElement(\"style\");\n    element.appendChild(document.createTextNode(css));\n    document.body.appendChild(element);\n  }\n\n  const js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-3.3.4.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-3.3.4.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-3.3.4.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-3.3.4.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-3.3.4.min.js\"];\n  const css_urls = [];\n\n  const inline_js = [    function(Bokeh) {\n      Bokeh.set_log_level(\"info\");\n    },\nfunction(Bokeh) {\n    }\n  ];\n\n  function run_inline_js() {\n    if (root.Bokeh !== undefined || force === true) {\n          for (let i = 0; i < inline_js.length; i++) {\n      inline_js[i].call(root, root.Bokeh);\n    }\nif (force === true) {\n        display_loaded();\n      }} else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(run_inline_js, 100);\n    } else if (!root._bokeh_failed_load) {\n      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n      root._bokeh_failed_load = true;\n    } else if (force !== true) {\n      const cell = $(document.getElementById(\"ea9b634a-6886-431d-a6f9-3cd0086eb6da\")).parents('.cell').data().cell;\n      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n    }\n  }\n\n  if (root._bokeh_is_loading === 0) {\n    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n    run_inline_js();\n  } else {\n    load_libs(css_urls, js_urls, function() {\n      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n      run_inline_js();\n    });\n  }\n}(window));",
      "application/vnd.bokehjs_load.v0+json": ""
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/dc16/gentropy/.venv/lib/python3.11/site-packages/tqdm/auto.py:21: TqdmWarning:\n",
      "\n",
      "IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import pyspark.sql.functions as f\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "\n",
    "from gentropy.common.session import Session\n",
    "from gentropy.dataset.colocalisation import Colocalisation\n",
    "from gentropy.dataset.l2g_prediction import L2GPrediction\n",
    "from gentropy.dataset.study_index import StudyIndex\n",
    "from gentropy.dataset.study_locus import StudyLocus\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting default log level to \"WARN\".\n",
      "To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).\n",
      "25/04/10 11:19:46 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable\n"
     ]
    }
   ],
   "source": [
    "session = Session(\n",
    "    extended_spark_conf={\n",
    "        'spark.executor.memory':'10g',\n",
    "        'spark.driver.memory':'10g'}\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "target = session.spark.read.parquet('/users/dc16/data/releases/25.03/target')\n",
    "l2g = L2GPrediction.from_parquet(\n",
    "    session, \"/users/dc16/data/releases/25.03/l2g_prediction\"\n",
    ").df.select(\"studyLocusId\", \"geneId\", \"score\")\n",
    "cred_sets = StudyLocus.from_parquet(\n",
    "    session, \"/users/dc16/data/releases/25.03/credible_set\"\n",
    ")\n",
    "studies = (\n",
    "    session.spark.read.parquet(\n",
    "        \"/users/dc16/data/gentropy_paper/gwas_study_index_therapeutic_areas\"\n",
    "    )\n",
    "    .select(\n",
    "        \"studyId\",\n",
    "        \"traitFromSource\",\n",
    "        \"traitFromSourceMappedIds\",\n",
    "        \"nCases\",\n",
    "        \"nControls\",\n",
    "        \"nSamples\",\n",
    "        \"Haematology\",\n",
    "        \"Metabolic\",\n",
    "        \"Congenital\",\n",
    "        \"Signs/symptoms\",\n",
    "        \"Neurology\",\n",
    "        \"Immune\",\n",
    "        \"Psychiatry\",\n",
    "        \"Dermatology\",\n",
    "        \"Ophthalmology\",\n",
    "        \"Cardiovascular\",\n",
    "        \"Oncology\",\n",
    "        \"Respiratory\",\n",
    "        \"Digestive\",\n",
    "        \"Endocrine\",\n",
    "        \"Musculoskeletal\",\n",
    "        \"Infection\",\n",
    "        \"Measurement\",\n",
    "        f.col(\"bianry_less_cases\").alias(\"binary_less_cases\"),\n",
    "        \"Other\",\n",
    "    )\n",
    "    .filter(f.col(\"Measurement\") == 0)\n",
    "    .filter(f.col(\"bianry_less_cases\") == 1)\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "l2g_signif = (\n",
    "    l2g.filter(f.col(\"score\") >= 0.5)\n",
    "    .join(\n",
    "        cred_sets.df.select(\n",
    "            \"studyLocusId\",\n",
    "            \"variantId\",\n",
    "            \"studyId\",\n",
    "            \"beta\",\n",
    "            \"pValueMantissa\",\n",
    "            \"pValueExponent\",\n",
    "        ),\n",
    "        \"studyLocusId\",\n",
    "        \"inner\",\n",
    "    )\n",
    "    .join(studies, \"studyId\", \"inner\")\n",
    ")\n",
    "l2g_margin = (\n",
    "    l2g.filter(f.col(\"score\") >= 0.05)\n",
    "    .join(\n",
    "        cred_sets.df.select(\n",
    "            \"studyLocusId\",\n",
    "            \"variantId\",\n",
    "            \"studyId\",\n",
    "            \"beta\",\n",
    "            \"pValueMantissa\",\n",
    "            \"pValueExponent\",\n",
    "        ),\n",
    "        \"studyLocusId\",\n",
    "        \"inner\",\n",
    "    )\n",
    "    .join(studies, \"studyId\", \"inner\")\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "25/04/10 11:19:50 WARN SparkStringUtils: Truncated the string representation of a plan since it was too large. This behavior can be adjusted by setting 'spark.sql.debug.maxToStringFields'.\n",
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+--------------------------------+--------------------------------+---------------+------------------+----------------+---------------------+--------------+--------------+---------------------------------------------------------------------------------------------------------------------+------------------------------+------+---------+--------+-----------+---------+----------+--------------+---------+------+----------+-----------+-------------+--------------+--------+-----------+---------+---------+---------------+---------+-----------+-----------------+-----+\n",
      "|studyId                         |studyLocusId                    |geneId         |score             |variantId       |beta                 |pValueMantissa|pValueExponent|traitFromSource                                                                                                      |traitFromSourceMappedIds      |nCases|nControls|nSamples|Haematology|Metabolic|Congenital|Signs/symptoms|Neurology|Immune|Psychiatry|Dermatology|Ophthalmology|Cardiovascular|Oncology|Respiratory|Digestive|Endocrine|Musculoskeletal|Infection|Measurement|binary_less_cases|Other|\n",
      "+--------------------------------+--------------------------------+---------------+------------------+----------------+---------------------+--------------+--------------+---------------------------------------------------------------------------------------------------------------------+------------------------------+------+---------+--------+-----------+---------+----------+--------------+---------+------+----------+-----------+-------------+--------------+--------+-----------+---------+---------+---------------+---------+-----------+-----------------+-----+\n",
      "|GCST90014444                    |00174f955f2635a265a28053055e4991|ENSG00000169914|0.7356408164401692|1_19872444_A_G  |-0.011877935642681822|7.9714103     |-12           |Inflammatory bowel disease                                                                                           |[EFO_0003767]                 |5105  |324074   |329179  |0          |0        |1         |0             |0        |1     |0         |0          |0            |0             |0       |0          |1        |0        |0              |0        |0          |1                |0    |\n",
      "|GCST90244660                    |001943e996d3762cf5305c7d1f40568d|ENSG00000075539|0.7148654102181217|4_48691230_A_T  |NULL                 |1.0           |-10           |Migraine or type 2 diabetes                                                                                          |[MONDO_0005148, MONDO_0005277]|176208|1595263  |1771471 |0          |1        |0         |0             |1        |0     |0         |0          |0            |1             |0       |0          |0        |1        |0              |0        |0          |1                |0    |\n",
      "|GCST90271317                    |002f457a4b7b8851b34ab79f480bd1fb|ENSG00000124203|0.7309521563473781|20_59141918_G_A |NULL                 |1.0           |-13           |Gestational hypertension                                                                                             |[EFO_0000668]                 |8961  |184925   |193886  |0          |0        |1         |0             |0        |0     |0         |0          |0            |1             |0       |0          |0        |0        |0              |0        |0          |1                |0    |\n",
      "|GCST90042696                    |0043f21a03f0ecf9ab0e9e089d8fd8c0|ENSG00000275410|0.7397703276200222|17_37743574_A_G |-0.10766274731032938 |4.8268576     |-9            |Illnesses of siblings: Prostate cancer (UKB data field 20111_13)                                                     |[EFO_0000493, MONDO_0008315]  |5873  |350300   |356173  |0          |0        |0         |0             |0        |0     |0         |0          |0            |0             |1       |0          |0        |0        |0              |0        |0          |1                |0    |\n",
      "|GCST010480                      |005fac203d2bbc82f1fe8fabc736cbad|ENSG00000142046|0.530008822768115 |19_41359388_A_G |-0.11332868530700324 |7.0           |-9            |Coronary artery disease                                                                                              |[EFO_0001645]                 |15302 |36140    |51442   |0          |0        |0         |0             |0        |0     |0         |0          |0            |1             |0       |0          |0        |0        |0              |0        |0          |1                |0    |\n",
      "|GCST90446792                    |0062121bed6417ad2f07608247808204|ENSG00000120337|0.5824511365998359|1_172877652_A_G |0.3803818514176848   |4.181496      |-24           |Crohn's disease                                                                                                      |[EFO_0000384]                 |20873 |346719   |367592  |0          |0        |1         |0             |0        |1     |0         |0          |0            |0             |0       |0          |1        |0        |0              |0        |0          |1                |0    |\n",
      "|FINNGEN_R12_C3_OTHER_SKIN_EXALLC|0099c32183de854cfaadbf88f7a3acc5|ENSG00000164175|0.7491775647319908|5_33957552_T_C  |-0.0584742310594291  |2.086         |-16           |Other malignant neoplasms of skin (=non-melanoma skin cancer), excluding all cancers (controls excluding all cancers)|[EFO_0009260]                 |28424 |378749   |407173  |0          |0        |0         |0             |0        |0     |0         |0          |0            |0             |1       |0          |0        |0        |0              |0        |0          |1                |0    |\n",
      "|GCST90134662                    |009a8bb56fd747194c2bacf91bb2e361|ENSG00000164362|0.6701172245813725|5_1287079_G_A   |-0.26                |2.0           |-110          |Lung adenocarcinoma                                                                                                  |[EFO_0000571]                 |15265 |32319    |47584   |0          |0        |0         |0             |0        |0     |0         |0          |0            |0             |1       |0          |0        |0        |0              |0        |0          |1                |0    |\n",
      "|GCST90018926                    |00c955ef571ceec93dd581fef7fd8192|ENSG00000118520|0.8990471249883547|6_131619344_T_C |0.007052114968332962 |6.871711      |-10           |Type 2 diabetes                                                                                                      |[MONDO_0005148]               |84224 |583280   |667504  |0          |1        |0         |0             |0        |0     |0         |0          |0            |0             |0       |0          |0        |1        |0              |0        |0          |1                |0    |\n",
      "|GCST90018926                    |00c955ef571ceec93dd581fef7fd8192|ENSG00000112282|0.7019896618865982|6_131619344_T_C |0.007052114968332962 |6.871711      |-10           |Type 2 diabetes                                                                                                      |[MONDO_0005148]               |84224 |583280   |667504  |0          |1        |0         |0             |0        |0     |0         |0          |0            |0             |0       |0          |0        |1        |0              |0        |0          |1                |0    |\n",
      "|GCST90018657                    |00db168aa29da9f7983f627ac598e903|ENSG00000187498|0.9101135132608806|13_110302163_T_C|0.017817303925689275 |6.025843      |-15           |Myocardial infarction                                                                                                |[EFO_0000612]                 |14992 |146214   |161206  |0          |0        |0         |0             |0        |0     |0         |0          |0            |1             |0       |0          |0        |0        |0              |0        |0          |1                |0    |\n",
      "|GCST90016591                    |00e70f85d2caba4651ed8cc2e1a52d52|ENSG00000133794|0.5056901649965588|11_13149238_G_T |-0.07257069283483537 |3.0           |-9            |Chronic obstructive pulmonary disease x current smoker interaction (main effect)                                     |[EFO_0000341]                 |21077 |179689   |200766  |0          |0        |0         |0             |0        |0     |0         |0          |0            |0             |0       |1          |0        |0        |0              |0        |0          |1                |0    |\n",
      "|GCST90044402                    |00f9ce9319f85ba9f8ed0e5c75f2cf16|ENSG00000142611|0.6903584599969158|1_3168622_C_T   |0.010757224020170566 |3.4171448     |-15           |Pain type(s) experienced in last month: Headache (UKB data field 6159_1)                                             |[HP_0002315]                  |91724 |363548   |455272  |0          |0        |0         |0             |1        |0     |0         |0          |0            |0             |0       |0          |0        |0        |0              |0        |0          |1                |0    |\n",
      "|GCST90454344                    |010e6a9f353d1d1d209b75f8a496165f|ENSG00000196588|0.5501994561128694|22_40654302_C_G |1.0006384363492555   |1.876902      |-9            |Breast cancer (triple negative)                                                                                      |[EFO_0005537]                 |14900 |91477    |106377  |0          |0        |0         |0             |0        |0     |0         |0          |0            |0             |1       |0          |0        |0        |0              |0        |0          |1                |0    |\n",
      "|GCST90018799                    |011321ebb26a508b03f6d905adae9b9a|ENSG00000103460|0.728363641709674 |16_52565276_C_T |0.07756001631680763  |3.865916      |-92           |Breast cancer                                                                                                        |[EFO_0000305]                 |23714 |313566   |337280  |0          |0        |0         |0             |0        |0     |0         |0          |0            |0             |1       |0          |0        |0        |0              |0        |0          |1                |0    |\n",
      "|GCST90319327                    |0121d77d77e7cb1a779c98f644f429b5|ENSG00000068796|0.5341768431886815|5_62217019_A_C  |0.0222967            |2.0           |-15           |Depression                                                                                                           |[MONDO_0002009]               |371184|978703   |1349887 |0          |0        |0         |0             |1        |0     |1         |0          |0            |0             |0       |0          |0        |0        |0              |0        |0          |1                |0    |\n",
      "|GCST90018802                    |012b3a5794d26ae32619999d298c459e|ENSG00000176165|0.52115339357377  |14_28545329_G_A |4.7477               |8.983         |-9            |Bronchitis                                                                                                           |[EFO_0009661]                 |5948  |644818   |650766  |0          |0        |0         |0             |0        |0     |0         |0          |0            |0             |0       |1          |0        |0        |0              |0        |0          |1                |0    |\n",
      "|GCST90104534                    |012fa62107fbc69a65b4605ab6f77a63|ENSG00000137273|0.540622074220468 |6_1366060_C_T   |-0.005413698211906131|7.2154565     |-10           |Stroke                                                                                                               |[EFO_0000712]                 |110182|1503898  |1614080 |0          |0        |0         |0             |1        |0     |0         |0          |0            |1             |0       |0          |0        |0        |0              |0        |0          |1                |0    |\n",
      "|GCST90271336                    |0159ad53d5ea67a6f1c4fa550b66f96c|ENSG00000167653|0.6877188055000317|8_142674302_C_A |NULL                 |7.0           |-10           |Peptic ulcer disease or anorexia nervosa (pleiotropy)                                                                |[MONDO_0004247, MONDO_0005351]|33658 |495186   |528844  |0          |1        |0         |0             |1        |0     |1         |1          |0            |0             |0       |0          |1        |0        |0              |0        |0          |1                |0    |\n",
      "|GCST005536                      |015e4ec32d268aa6ae4aa4111ef784a6|ENSG00000184730|0.9314255030338043|16_28495107_C_T |0.17100779689368129  |4.803         |-11           |Type 1 diabetes                                                                                                      |[MONDO_0005147]               |6683  |12173    |21526   |0          |1        |0         |0             |0        |1     |0         |0          |0            |0             |0       |0          |1        |1        |0              |0        |0          |1                |0    |\n",
      "+--------------------------------+--------------------------------+---------------+------------------+----------------+---------------------+--------------+--------------+---------------------------------------------------------------------------------------------------------------------+------------------------------+------+---------+--------+-----------+---------+----------+--------------+---------+------+----------+-----------+-------------+--------------+--------+-----------+---------+---------+---------------+---------+-----------+-----------------+-----+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "l2g_signif.show(truncate=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 16:======>                                                   (1 + 8) / 9]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-RECORD 0---------------\n",
      " Haematology     | 394  \n",
      " Metabolic       | 5861 \n",
      " Congenital      | 5480 \n",
      " Signs/symptoms  | 265  \n",
      " Neurology       | 6677 \n",
      " Immune          | 6416 \n",
      " Psychiatry      | 4039 \n",
      " Dermatology     | 1535 \n",
      " Ophthalmology   | 2628 \n",
      " Cardiovascular  | 7205 \n",
      " Oncology        | 7176 \n",
      " Respiratory     | 3167 \n",
      " Digestive       | 3703 \n",
      " Endocrine       | 5644 \n",
      " Musculoskeletal | 5287 \n",
      " Infection       | 719  \n",
      " Other           | 2919 \n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "(\n",
    "    l2g_signif\n",
    "    .agg(\n",
    "        f.sum('Haematology').alias('Haematology'),\n",
    "        f.sum('Metabolic').alias('Metabolic'),\n",
    "        f.sum('Congenital').alias('Congenital'),\n",
    "        f.sum('Signs/symptoms').alias('Signs/symptoms'),\n",
    "        f.sum('Neurology').alias('Neurology'),\n",
    "        f.sum('Immune').alias('Immune'),\n",
    "        f.sum('Psychiatry').alias('Psychiatry'),\n",
    "        f.sum('Dermatology').alias('Dermatology'),\n",
    "        f.sum('Ophthalmology').alias('Ophthalmology'),\n",
    "        f.sum('Cardiovascular').alias('Cardiovascular'),\n",
    "        f.sum('Oncology').alias('Oncology'),\n",
    "        f.sum('Respiratory').alias('Respiratory'),\n",
    "        f.sum('Digestive').alias('Digestive'),\n",
    "        f.sum('Endocrine').alias('Endocrine'),\n",
    "        f.sum('Musculoskeletal').alias('Musculoskeletal'),\n",
    "        f.sum('Infection').alias('Infection'),\n",
    "        f.sum('Other').alias('Other'),\n",
    "    )\n",
    "    .show(vertical=True)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 23:=================================================>      (23 + 3) / 26]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-RECORD 0----------------\n",
      " Haematology     | 1061  \n",
      " Metabolic       | 13174 \n",
      " Congenital      | 13071 \n",
      " Signs/symptoms  | 760   \n",
      " Neurology       | 16564 \n",
      " Immune          | 15445 \n",
      " Psychiatry      | 10282 \n",
      " Dermatology     | 3793  \n",
      " Ophthalmology   | 5615  \n",
      " Cardiovascular  | 16869 \n",
      " Oncology        | 15381 \n",
      " Respiratory     | 7788  \n",
      " Digestive       | 9113  \n",
      " Endocrine       | 12471 \n",
      " Musculoskeletal | 12803 \n",
      " Infection       | 1683  \n",
      " Other           | 6724  \n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "(\n",
    "    l2g_margin\n",
    "    .agg(\n",
    "        f.sum('Haematology').alias('Haematology'),\n",
    "        f.sum('Metabolic').alias('Metabolic'),\n",
    "        f.sum('Congenital').alias('Congenital'),\n",
    "        f.sum('Signs/symptoms').alias('Signs/symptoms'),\n",
    "        f.sum('Neurology').alias('Neurology'),\n",
    "        f.sum('Immune').alias('Immune'),\n",
    "        f.sum('Psychiatry').alias('Psychiatry'),\n",
    "        f.sum('Dermatology').alias('Dermatology'),\n",
    "        f.sum('Ophthalmology').alias('Ophthalmology'),\n",
    "        f.sum('Cardiovascular').alias('Cardiovascular'),\n",
    "        f.sum('Oncology').alias('Oncology'),\n",
    "        f.sum('Respiratory').alias('Respiratory'),\n",
    "        f.sum('Digestive').alias('Digestive'),\n",
    "        f.sum('Endocrine').alias('Endocrine'),\n",
    "        f.sum('Musculoskeletal').alias('Musculoskeletal'),\n",
    "        f.sum('Infection').alias('Infection'),\n",
    "        f.sum('Other').alias('Other'),\n",
    "    )\n",
    "    .show(vertical=True)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "studies_per_therapeutic_area = (\n",
    "    studies\n",
    "    .filter(f.col(\"Measurement\") == 0)\n",
    "    .filter(f.col(\"bianry_less_cases\") == 1)\n",
    "    .agg(\n",
    "        f.sum('Haematology').alias('Haematology'),\n",
    "        f.sum('Metabolic').alias('Metabolic'),\n",
    "        f.sum('Congenital').alias('Congenital'),\n",
    "        f.sum('Signs/symptoms').alias('Signs/symptoms'),\n",
    "        f.sum('Neurology').alias('Neurology'),\n",
    "        f.sum('Immune').alias('Immune'),\n",
    "        f.sum('Psychiatry').alias('Psychiatry'),\n",
    "        f.sum('Dermatology').alias('Dermatology'),\n",
    "        f.sum('Ophthalmology').alias('Ophthalmology'),\n",
    "        f.sum('Cardiovascular').alias('Cardiovascular'),\n",
    "        f.sum('Oncology').alias('Oncology'),\n",
    "        f.sum('Respiratory').alias('Respiratory'),\n",
    "        f.sum('Digestive').alias('Digestive'),\n",
    "        f.sum('Endocrine').alias('Endocrine'),\n",
    "        f.sum('Musculoskeletal').alias('Musculoskeletal'),\n",
    "        f.sum('Infection').alias('Infection'),\n",
    "        f.sum('Other').alias('Other'),\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_therapeutic_areas = (\n",
    "    l2g_signif.groupBy(\"geneId\")\n",
    "    .agg(\n",
    "        f.sum(\"Haematology\").alias(\"Haematology\"),\n",
    "        f.sum(\"Metabolic\").alias(\"Metabolic\"),\n",
    "        f.sum(\"Congenital\").alias(\"Congenital\"),\n",
    "        f.sum(\"Signs/symptoms\").alias(\"Signs/symptoms\"),\n",
    "        f.sum(\"Neurology\").alias(\"Neurology\"),\n",
    "        f.sum(\"Immune\").alias(\"Immune\"),\n",
    "        f.sum(\"Psychiatry\").alias(\"Psychiatry\"),\n",
    "        f.sum(\"Dermatology\").alias(\"Dermatology\"),\n",
    "        f.sum(\"Ophthalmology\").alias(\"Ophthalmology\"),\n",
    "        f.sum(\"Cardiovascular\").alias(\"Cardiovascular\"),\n",
    "        f.sum(\"Oncology\").alias(\"Oncology\"),\n",
    "        f.sum(\"Respiratory\").alias(\"Respiratory\"),\n",
    "        f.sum(\"Digestive\").alias(\"Digestive\"),\n",
    "        f.sum(\"Endocrine\").alias(\"Endocrine\"),\n",
    "        f.sum(\"Musculoskeletal\").alias(\"Musculoskeletal\"),\n",
    "        f.sum(\"Infection\").alias(\"Infection\"),\n",
    "        f.sum(\"Other\").alias(\"Other\"),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"total\",\n",
    "        f.col(\"Haematology\")\n",
    "        + f.col(\"Metabolic\")\n",
    "        + f.col(\"Congenital\")\n",
    "        + f.col(\"Signs/symptoms\")\n",
    "        + f.col(\"Neurology\")\n",
    "        + f.col(\"Immune\")\n",
    "        + f.col(\"Psychiatry\")\n",
    "        + f.col(\"Dermatology\")\n",
    "        + f.col(\"Ophthalmology\")\n",
    "        + f.col(\"Cardiovascular\")\n",
    "        + f.col(\"Oncology\")\n",
    "        + f.col(\"Respiratory\")\n",
    "        + f.col(\"Digestive\")\n",
    "        + f.col(\"Endocrine\")\n",
    "        + f.col(\"Musculoskeletal\")\n",
    "        + f.col(\"Infection\")\n",
    "        + f.col(\"Other\"),\n",
    "    )\n",
    "    .withColumns(\n",
    "        {\n",
    "            'haematology_proportion': f.col('Haematology') / studies_per_therapeutic_area.collect()[0]['Haematology'],\n",
    "            'metabolic_proportion': f.col('Metabolic') / studies_per_therapeutic_area.collect()[0]['Metabolic'],\n",
    "            'congenital_proportion': f.col('Congenital') / studies_per_therapeutic_area.collect()[0]['Congenital'],\n",
    "            'signs_symptoms_proportion': f.col('Signs/symptoms') / studies_per_therapeutic_area.collect()[0]['Signs/symptoms'],\n",
    "            'neurology_proportion': f.col('Neurology') / studies_per_therapeutic_area.collect()[0]['Neurology'],\n",
    "            'immune_proportion': f.col('Immune') / studies_per_therapeutic_area.collect()[0]['Immune'],\n",
    "            'psychiatry_proportion': f.col('Psychiatry') / studies_per_therapeutic_area.collect()[0]['Psychiatry'],\n",
    "            'dermatology_proportion': f.col('Dermatology') / studies_per_therapeutic_area.collect()[0]['Dermatology'],\n",
    "            'ophthalmology_proportion': f.col('Ophthalmology') / studies_per_therapeutic_area.collect()[0]['Ophthalmology'],\n",
    "            'cardiovascular_proportion': f.col('Cardiovascular') / studies_per_therapeutic_area.collect()[0]['Cardiovascular'],\n",
    "            'oncology_proportion': f.col('Oncology') / studies_per_therapeutic_area.collect()[0]['Oncology'],\n",
    "            'respiratory_proportion': f.col('Respiratory') / studies_per_therapeutic_area.collect()[0]['Respiratory'],\n",
    "            'digestive_proportion': f.col('Digestive') / studies_per_therapeutic_area.collect()[0]['Digestive'],\n",
    "            'endocrine_proportion': f.col('Endocrine') / studies_per_therapeutic_area.collect()[0]['Endocrine'],\n",
    "            'musculoskeletal_proportion': f.col('Musculoskeletal') / studies_per_therapeutic_area.collect()[0]['Musculoskeletal'],\n",
    "            'infection_proportion': f.col('Infection') / studies_per_therapeutic_area.collect()[0]['Infection'],\n",
    "            'other_proportion': f.col('Other') / studies_per_therapeutic_area.collect()[0]['Other'],\n",
    "        }\n",
    "    )\n",
    "    .join(\n",
    "        target.withColumn(\"constraint\", f.explode(\"constraint\"))\n",
    "        .select(\n",
    "            \"id\",\n",
    "            \"approvedSymbol\",\n",
    "            \"biotype\",\n",
    "            \"genomicLocation.chromosome\",\n",
    "            \"genomicLocation.start\",\n",
    "            \"genomicLocation.end\",\n",
    "            \"constraint.constraintType\",\n",
    "            \"constraint.score\",\n",
    "        )\n",
    "        .filter(f.col(\"constraintType\") == \"lof\"),\n",
    "        target[\"id\"] == f.col(\"geneId\"),\n",
    "        \"inner\",\n",
    "    )\n",
    "    .drop(\"id\", \"constraintType\")\n",
    "    .withColumnRenamed(\"score\", \"lofConstraint\")\n",
    "    .withColumn(\n",
    "        'pleiotropy',\n",
    "        (\n",
    "            f.when(f.col(\"Haematology\") > 0, 1).otherwise(0)\n",
    "            + f.when(f.col(\"Metabolic\") > 0, 1).otherwise(0)\n",
    "            + f.when(f.col(\"Congenital\") > 0, 1).otherwise(0)\n",
    "            + f.when(f.col(\"Signs/symptoms\") > 0, 1).otherwise(0)\n",
    "            + f.when(f.col(\"Neurology\") > 0, 1).otherwise(0)\n",
    "            + f.when(f.col(\"Immune\") > 0, 1).otherwise(0)\n",
    "            + f.when(f.col(\"Psychiatry\") > 0, 1).otherwise(0)\n",
    "            + f.when(f.col(\"Dermatology\") > 0, 1).otherwise(0)\n",
    "            + f.when(f.col(\"Ophthalmology\") > 0, 1).otherwise(0)\n",
    "            + f.when(f.col(\"Cardiovascular\") > 0, 1).otherwise(0)\n",
    "            + f.when(f.col(\"Oncology\") > 0, 1).otherwise(0)\n",
    "            + f.when(f.col(\"Respiratory\") > 0, 1).otherwise(0)\n",
    "            + f.when(f.col(\"Digestive\") > 0, 1).otherwise(0)\n",
    "            + f.when(f.col(\"Endocrine\") > 0, 1).otherwise(0)\n",
    "            + f.when(f.col(\"Musculoskeletal\") > 0, 1).otherwise(0)\n",
    "            + f.when(f.col(\"Infection\") > 0, 1).otherwise(0)\n",
    "            + f.when(f.col(\"Other\") > 0, 1).otherwise(0)\n",
    "        )\n",
    "        / 17\n",
    "    )\n",
    "    .sort(f.desc(\"total\"))\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1698"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "studies_per_therapeutic_area.collect()[0]['Cardiovascular']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 197:============>                                            (2 + 7) / 9]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+-----------+---------+----------+--------------+---------+------+----------+-----------+-------------+--------------+--------+-----------+---------+---------+---------------+---------+-----+-----+----------------------+--------------------+---------------------+-------------------------+--------------------+--------------------+---------------------+----------------------+------------------------+-------------------------+--------------------+----------------------+--------------------+--------------------+--------------------------+--------------------+--------------------+--------------+--------------+----------+---------+---------+-------------+-------------------+\n",
      "|         geneId|Haematology|Metabolic|Congenital|Signs/symptoms|Neurology|Immune|Psychiatry|Dermatology|Ophthalmology|Cardiovascular|Oncology|Respiratory|Digestive|Endocrine|Musculoskeletal|Infection|Other|total|haematology_proportion|metabolic_proportion|congenital_proportion|signs_symptoms_proportion|neurology_proportion|   immune_proportion|psychiatry_proportion|dermatology_proportion|ophthalmology_proportion|cardiovascular_proportion| oncology_proportion|respiratory_proportion|digestive_proportion|endocrine_proportion|musculoskeletal_proportion|infection_proportion|    other_proportion|approvedSymbol|       biotype|chromosome|    start|      end|lofConstraint|         pleiotropy|\n",
      "+---------------+-----------+---------+----------+--------------+---------+------+----------+-----------+-------------+--------------+--------+-----------+---------+---------+---------------+---------+-----+-----+----------------------+--------------------+---------------------+-------------------------+--------------------+--------------------+---------------------+----------------------+------------------------+-------------------------+--------------------+----------------------+--------------------+--------------------+--------------------------+--------------------+--------------------+--------------+--------------+----------+---------+---------+-------------+-------------------+\n",
      "|ENSG00000147883|          0|       73|         5|             0|       38|     2|         4|          1|           49|           130|      70|          4|        2|       72|              0|        0|   19|  469|                   0.0| 0.10625909752547306| 0.004191114836546521|                      0.0|0.014147431124348473|0.001699235344095...| 0.003333333333333...|  0.001386962552011...|     0.05785123966942149|      0.07656065959952886| 0.03143242029636282|  0.004932182490752158|0.001607717041800643| 0.08955223880597014|                       0.0|                 0.0|0.006369426751592357|        CDKN2B|protein_coding|         9| 22002903| 22009305|    0.0073369| 0.7647058823529411|\n",
      "|ENSG00000130203|          0|       33|        29|             4|       77|     1|        60|          1|           12|            61|       1|          9|        8|       12|              9|        8|    9|  334|                   0.0|0.048034934497816595| 0.024308466051969825|      0.01020408163265306| 0.02866716306775875|8.496176720475786E-4|                 0.05|  0.001386962552011...|    0.014167650531286895|        0.035924617196702| 4.49034575662326E-4|  0.011097410604192354|0.006430868167202572|0.014925373134328358|      0.005056179775280...| 0.01199400299850075|0.003017096882333...|          APOE|protein_coding|        19| 44905791| 44909393|    0.0018685| 0.9411764705882353|\n",
      "|ENSG00000164093|          0|        0|        26|             0|       38|    24|         2|          0|            9|           173|       1|          5|       24|        0|              7|        0|    8|  317|                   0.0|                 0.0|  0.02179379715004191|                      0.0|0.014147431124348473|0.020390824129141887| 0.001666666666666...|                   0.0|    0.010625737898465172|      0.10188457008244994| 4.49034575662326E-4|  0.006165228113440197| 0.01929260450160772|                 0.0|      0.003932584269662922|                 0.0|0.002681863895407...|         PITX2|protein_coding|         4|110617423|110642123|      0.97618| 0.6470588235294118|\n",
      "|ENSG00000134242|          3|       32|        13|             0|       11|    93|         0|          0|            7|             5|      10|          0|       26|       50|             51|        0|    3|  304|   0.01507537688442211|0.046579330422125184| 0.010896898575020955|                      0.0|0.004095309009679821|  0.0790144435004248|                  0.0|                   0.0|    0.008264462809917356|     0.002944640753828033| 0.00449034575662326|                   0.0| 0.02090032154340836| 0.06218905472636816|      0.028651685393258425|                 0.0|0.001005698960777...|        PTPN22|protein_coding|         1|113813811|113871753|   3.4921E-21| 0.7058823529411765|\n",
      "|ENSG00000162594|          0|        2|        82|             0|        2|    92|         0|         13|            3|             0|       1|          0|       72|        2|             11|        2|    0|  282|                   0.0|0.002911208151382824|  0.06873428331936295|                      0.0|7.446016381236039E-4| 0.07816482582837724|                  0.0|  0.018030513176144243|    0.003541912632821...|                      0.0| 4.49034575662326E-4|                   0.0| 0.05787781350482315|0.002487562189054...|      0.006179775280898876|0.002998500749625...|                 0.0|         IL23R|protein_coding|         1| 67138907| 67259979|    5.5961E-4| 0.6470588235294118|\n",
      "|ENSG00000138378|          1|        1|        21|             0|        4|    93|         0|         17|           20|             2|       6|         15|       10|       13|             72|        2|    4|  281|  0.005025125628140704|0.001455604075691412|  0.01760268231349539|                      0.0|0.001489203276247...|  0.0790144435004248|                  0.0|  0.023578363384188627|    0.023612750885478158|     0.001177856301531213|0.002694207453973956|  0.018495684340320593|0.008038585209003215| 0.01616915422885572|       0.04044943820224719|0.002998500749625...|0.001340931947703...|         STAT4|protein_coding|         2|191029576|191178435|       0.7662| 0.8823529411764706|\n",
      "|ENSG00000148737|          0|      102|         2|             0|        6|     4|         3|          2|            6|             8|      15|          0|        2|       99|              2|        0|    4|  255|                   0.0| 0.14847161572052403| 0.001676445934618...|                      0.0|0.002233804914370...|0.003398470688190...|               0.0025|  0.002773925104022...|    0.007083825265643...|     0.004711425206124852| 0.00673551863493489|                   0.0|0.001607717041800643| 0.12313432835820895|      0.001123595505617...|                 0.0|0.001340931947703...|        TCF7L2|protein_coding|        10|112950247|113167678|      0.99537| 0.7647058823529411|\n",
      "|ENSG00000212993|          0|        0|         1|             0|        1|     9|         0|          0|            0|             0|     210|          0|        0|       13|              0|        0|    1|  235|                   0.0|                 0.0| 8.382229673093043E-4|                      0.0|3.723008190618019...|0.007646559048428...|                  0.0|                   0.0|                     0.0|                      0.0| 0.09429726088908846|                   0.0|                 0.0| 0.01616915422885572|                       0.0|                 0.0|3.352329869259135...|       POU5F1B|protein_coding|         8|127322183|127420066|    2.9118E-4|0.35294117647058826|\n",
      "|ENSG00000140718|          1|       58|        12|             0|       14|     6|        11|          3|            0|            21|      18|         12|        3|       36|             24|        1|    6|  226|  0.005025125628140704| 0.08442503639010189| 0.010058675607711651|                      0.0|0.005212211466865...|0.005097706032285472| 0.009166666666666667|  0.004160887656033287|                     0.0|     0.012367491166077738|0.008082622361921869|  0.014796547472256474|0.002411575562700965| 0.04477611940298507|       0.01348314606741573|0.001499250374812...|0.002011397921555481|           FTO|protein_coding|        16| 53701692| 54158512|     8.364E-8| 0.8823529411764706|\n",
      "|ENSG00000105397|          0|        9|        27|             0|        1|    70|         1|         19|            2|             0|       3|          7|       15|       14|             34|       23|    0|  225|                   0.0|0.013100436681222707| 0.022632020117351215|                      0.0|3.723008190618019...|0.059473237043330504| 8.333333333333334E-4|  0.026352288488210817|    0.002361275088547...|                      0.0|0.001347103726986978|  0.008631319358816275|0.012057877813504822|0.017412935323383085|      0.019101123595505618|0.034482758620689655|                 0.0|          TYK2|protein_coding|        19| 10350528| 10380608|    4.4963E-8| 0.7647058823529411|\n",
      "|ENSG00000164362|          1|        0|         5|             0|        0|     1|         0|          5|            0|             4|     190|          7|        1|        2|              1|        0|    6|  223|  0.005025125628140704|                 0.0| 0.004191114836546521|                      0.0|                 0.0|8.496176720475786E-4|                  0.0|  0.006934812760055479|                     0.0|     0.002355712603062426| 0.08531656937584194|  0.008631319358816275|8.038585209003215E-4|0.002487562189054...|      5.617977528089888E-4|                 0.0|0.002011397921555481|          TERT|protein_coding|         5|  1253147|  1295068|      0.99046| 0.6470588235294118|\n",
      "|ENSG00000145996|          1|       65|        18|             0|        1|    20|         0|          5|            0|             4|      16|          0|       14|       65|              0|        0|    2|  211|  0.005025125628140704| 0.09461426491994178| 0.015088013411567477|                      0.0|3.723008190618019...|0.016992353440951572|                  0.0|  0.006934812760055479|                     0.0|     0.002355712603062426|0.007184553210597216|                   0.0|0.011254019292604502| 0.08084577114427861|                       0.0|                 0.0|6.704659738518271E-4|        CDKAL1|protein_coding|         6| 20534457| 21232404|     0.016021| 0.6470588235294118|\n",
      "|ENSG00000145012|          0|       11|        20|             1|        0|    36|         0|          6|            8|             0|      23|         44|       17|       26|              4|        0|    4|  200|                   0.0| 0.01601164483260553| 0.016764459346186086|     0.002551020408163265|                 0.0| 0.03058623619371283|                  0.0|  0.008321775312066574|    0.009445100354191263|                      0.0|0.010327795240233499|   0.05425400739827373|0.013665594855305467| 0.03233830845771144|      0.002247191011235...|                 0.0|0.001340931947703...|           LPP|protein_coding|         3|188153284|188890671|    0.0071908| 0.7058823529411765|\n",
      "|ENSG00000134954|          1|       20|        12|             0|        2|    48|         1|         15|            9|             4|       4|         13|        5|       20|             27|        0|    3|  184|  0.005025125628140704| 0.02911208151382824| 0.010058675607711651|                      0.0|7.446016381236039E-4| 0.04078164825828377| 8.333333333333334E-4|  0.020804438280166437|    0.010625737898465172|     0.002355712603062426|0.001796138302649304|  0.016029593094944512|0.004019292604501...|0.024875621890547265|      0.015168539325842697|                 0.0|0.001005698960777...|          ETS1|protein_coding|        11|128458761|128587558|       0.7816| 0.8823529411764706|\n",
      "|ENSG00000166949|          0|        0|        19|             3|        1|    29|         0|          2|            2|            11|      10|         63|       15|        1|             19|        1|    7|  183|                   0.0|                 0.0|  0.01592623637887678|     0.007653061224489796|3.723008190618019...| 0.02463891248937978|                  0.0|  0.002773925104022...|    0.002361275088547...|     0.006478209658421673| 0.00449034575662326|   0.07768187422934648|0.012057877813504822|0.001243781094527...|      0.010674157303370787|0.001499250374812...|0.002346630908481...|         SMAD3|protein_coding|        15| 67063763| 67195173|      0.79796| 0.8235294117647058|\n",
      "|ENSG00000111252|          1|       14|        13|             0|        8|    36|         1|          3|            3|            28|       9|          6|       20|       23|             14|        0|    2|  181|  0.005025125628140704|0.020378457059679767| 0.010896898575020955|                      0.0|0.002978406552494...| 0.03058623619371283| 8.333333333333334E-4|  0.004160887656033287|    0.003541912632821...|     0.016489988221436984|0.004041311180960934|  0.007398273736128237| 0.01607717041800643|0.028606965174129355|      0.007865168539325843|                 0.0|6.704659738518271E-4|         SH2B3|protein_coding|        12|111405923|111451623|    1.0289E-4| 0.8823529411764706|\n",
      "|ENSG00000134460|          0|       12|        11|             0|        3|    49|         0|         12|            0|             0|       3|         23|       21|       25|             15|        0|    4|  178|                   0.0|0.017467248908296942| 0.009220452640402346|                      0.0|0.001116902457185...|0.041631265930331354|                  0.0|  0.016643550624133148|                     0.0|                      0.0|0.001347103726986978|   0.02836004932182491|0.016881028938906754| 0.03109452736318408|      0.008426966292134831|                 0.0|0.001340931947703...|         IL2RA|protein_coding|        10|  6010689|  6062370|     0.054709| 0.6470588235294118|\n",
      "|ENSG00000112182|          0|        9|        11|             1|        1|    38|         0|          4|            0|             0|      18|         46|       16|       27|              5|        1|    1|  178|                   0.0|0.013100436681222707| 0.009220452640402346|     0.002551020408163265|3.723008190618019...| 0.03228547153780799|                  0.0|  0.005547850208044383|                     0.0|                      0.0|0.008082622361921869|   0.05672009864364982|0.012861736334405145|0.033582089552238806|      0.002808988764044...|0.001499250374812...|3.352329869259135...|         BACH2|protein_coding|         6| 89926528| 90296843|      0.99803| 0.7647058823529411|\n",
      "|ENSG00000164512|          0|       40|         6|             0|        5|    28|         0|          0|           16|             5|       1|          0|        5|       37|             23|        0|    6|  172|                   0.0| 0.05822416302765648| 0.005029337803855826|                      0.0|0.001861504095309...|  0.0237892948173322|                  0.0|                   0.0|    0.018890200708382526|     0.002944640753828033| 4.49034575662326E-4|                   0.0|0.004019292604501...|0.046019900497512436|      0.012921348314606741|                 0.0|0.002011397921555481|       ANKRD55|protein_coding|         5| 56099680| 56233330|    6.7577E-8| 0.6470588235294118|\n",
      "|ENSG00000171522|          0|        1|        44|             0|        4|    53|         0|          5|            0|             0|      11|          4|       46|        1|              1|        0|    2|  172|                   0.0|0.001455604075691412| 0.036881810561609385|                      0.0|0.001489203276247...|0.045029736618521665|                  0.0|  0.006934812760055479|                     0.0|                      0.0|0.004939380332285586|  0.004932182490752158| 0.03697749196141479|0.001243781094527...|      5.617977528089888E-4|                 0.0|6.704659738518271E-4|        PTGER4|protein_coding|         5| 40679915| 40693735|       0.9755| 0.6470588235294118|\n",
      "+---------------+-----------+---------+----------+--------------+---------+------+----------+-----------+-------------+--------------+--------+-----------+---------+---------+---------------+---------+-----+-----+----------------------+--------------------+---------------------+-------------------------+--------------------+--------------------+---------------------+----------------------+------------------------+-------------------------+--------------------+----------------------+--------------------+--------------------+--------------------------+--------------------+--------------------+--------------+--------------+----------+---------+---------+-------------+-------------------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "genes_therapeutic_areas.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "genes_pd = genes_therapeutic_areas.toPandas()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneId</th>\n",
       "      <th>Haematology</th>\n",
       "      <th>Metabolic</th>\n",
       "      <th>Congenital</th>\n",
       "      <th>Signs/symptoms</th>\n",
       "      <th>Neurology</th>\n",
       "      <th>Immune</th>\n",
       "      <th>Psychiatry</th>\n",
       "      <th>Dermatology</th>\n",
       "      <th>Ophthalmology</th>\n",
       "      <th>...</th>\n",
       "      <th>Musculoskeletal</th>\n",
       "      <th>Infection</th>\n",
       "      <th>Other</th>\n",
       "      <th>total</th>\n",
       "      <th>approvedSymbol</th>\n",
       "      <th>biotype</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>lofScore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000147883</td>\n",
       "      <td>0</td>\n",
       "      <td>73</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>38</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>49</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>19</td>\n",
       "      <td>469</td>\n",
       "      <td>CDKN2B</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>9</td>\n",
       "      <td>22002903</td>\n",
       "      <td>22009305</td>\n",
       "      <td>7.336900e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000130203</td>\n",
       "      <td>0</td>\n",
       "      <td>33</td>\n",
       "      <td>29</td>\n",
       "      <td>4</td>\n",
       "      <td>77</td>\n",
       "      <td>1</td>\n",
       "      <td>60</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>...</td>\n",
       "      <td>9</td>\n",
       "      <td>8</td>\n",
       "      <td>9</td>\n",
       "      <td>334</td>\n",
       "      <td>APOE</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>19</td>\n",
       "      <td>44905791</td>\n",
       "      <td>44909393</td>\n",
       "      <td>1.868500e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000164093</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>26</td>\n",
       "      <td>0</td>\n",
       "      <td>38</td>\n",
       "      <td>24</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>...</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>317</td>\n",
       "      <td>PITX2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>4</td>\n",
       "      <td>110617423</td>\n",
       "      <td>110642123</td>\n",
       "      <td>9.761800e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000134242</td>\n",
       "      <td>3</td>\n",
       "      <td>32</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>93</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>...</td>\n",
       "      <td>51</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>304</td>\n",
       "      <td>PTPN22</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1</td>\n",
       "      <td>113813811</td>\n",
       "      <td>113871753</td>\n",
       "      <td>3.492100e-21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000162594</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>82</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>92</td>\n",
       "      <td>0</td>\n",
       "      <td>13</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>11</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>282</td>\n",
       "      <td>IL23R</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1</td>\n",
       "      <td>67138907</td>\n",
       "      <td>67259979</td>\n",
       "      <td>5.596100e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6829</th>\n",
       "      <td>ENSG00000111145</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>ELK3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>12</td>\n",
       "      <td>96194375</td>\n",
       "      <td>96269824</td>\n",
       "      <td>8.350800e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6830</th>\n",
       "      <td>ENSG00000141905</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>NFIC</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>19</td>\n",
       "      <td>3314403</td>\n",
       "      <td>3469217</td>\n",
       "      <td>9.053400e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6831</th>\n",
       "      <td>ENSG00000177189</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>RPS6KA3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>X</td>\n",
       "      <td>20149911</td>\n",
       "      <td>20267519</td>\n",
       "      <td>9.999900e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6832</th>\n",
       "      <td>ENSG00000123094</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>RASSF8</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>12</td>\n",
       "      <td>25958232</td>\n",
       "      <td>26079892</td>\n",
       "      <td>9.415800e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6833</th>\n",
       "      <td>ENSG00000025708</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>TYMP</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>22</td>\n",
       "      <td>50525752</td>\n",
       "      <td>50530032</td>\n",
       "      <td>1.169400e-04</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6834 rows  25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               geneId  Haematology  Metabolic  Congenital  Signs/symptoms  \\\n",
       "0     ENSG00000147883            0         73           5               0   \n",
       "1     ENSG00000130203            0         33          29               4   \n",
       "2     ENSG00000164093            0          0          26               0   \n",
       "3     ENSG00000134242            3         32          13               0   \n",
       "4     ENSG00000162594            0          2          82               0   \n",
       "...               ...          ...        ...         ...             ...   \n",
       "6829  ENSG00000111145            0          0           0               0   \n",
       "6830  ENSG00000141905            0          0           0               0   \n",
       "6831  ENSG00000177189            0          0           0               0   \n",
       "6832  ENSG00000123094            0          0           0               0   \n",
       "6833  ENSG00000025708            0          0           0               0   \n",
       "\n",
       "      Neurology  Immune  Psychiatry  Dermatology  Ophthalmology  ...  \\\n",
       "0            38       2           4            1             49  ...   \n",
       "1            77       1          60            1             12  ...   \n",
       "2            38      24           2            0              9  ...   \n",
       "3            11      93           0            0              7  ...   \n",
       "4             2      92           0           13              3  ...   \n",
       "...         ...     ...         ...          ...            ...  ...   \n",
       "6829          0       0           0            0              0  ...   \n",
       "6830          0       0           0            0              0  ...   \n",
       "6831          0       0           0            0              0  ...   \n",
       "6832          0       0           0            0              0  ...   \n",
       "6833          0       0           0            0              0  ...   \n",
       "\n",
       "      Musculoskeletal  Infection  Other  total  approvedSymbol  \\\n",
       "0                   0          0     19    469          CDKN2B   \n",
       "1                   9          8      9    334            APOE   \n",
       "2                   7          0      8    317           PITX2   \n",
       "3                  51          0      3    304          PTPN22   \n",
       "4                  11          2      0    282           IL23R   \n",
       "...               ...        ...    ...    ...             ...   \n",
       "6829                0          0      0      1            ELK3   \n",
       "6830                0          0      0      1            NFIC   \n",
       "6831                0          0      0      1         RPS6KA3   \n",
       "6832                0          0      0      1          RASSF8   \n",
       "6833                0          0      0      1            TYMP   \n",
       "\n",
       "             biotype  chromosome      start        end      lofScore  \n",
       "0     protein_coding           9   22002903   22009305  7.336900e-03  \n",
       "1     protein_coding          19   44905791   44909393  1.868500e-03  \n",
       "2     protein_coding           4  110617423  110642123  9.761800e-01  \n",
       "3     protein_coding           1  113813811  113871753  3.492100e-21  \n",
       "4     protein_coding           1   67138907   67259979  5.596100e-04  \n",
       "...              ...         ...        ...        ...           ...  \n",
       "6829  protein_coding          12   96194375   96269824  8.350800e-02  \n",
       "6830  protein_coding          19    3314403    3469217  9.053400e-02  \n",
       "6831  protein_coding           X   20149911   20267519  9.999900e-01  \n",
       "6832  protein_coding          12   25958232   26079892  9.415800e-01  \n",
       "6833  protein_coding          22   50525752   50530032  1.169400e-04  \n",
       "\n",
       "[6834 rows x 25 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "root\n",
      " |-- id: string (nullable = true)\n",
      " |-- approvedSymbol: string (nullable = true)\n",
      " |-- biotype: string (nullable = true)\n",
      " |-- transcriptIds: array (nullable = true)\n",
      " |    |-- element: string (containsNull = true)\n",
      " |-- canonicalTranscript: struct (nullable = true)\n",
      " |    |-- id: string (nullable = true)\n",
      " |    |-- chromosome: string (nullable = true)\n",
      " |    |-- start: long (nullable = true)\n",
      " |    |-- end: long (nullable = true)\n",
      " |    |-- strand: string (nullable = true)\n",
      " |-- canonicalExons: array (nullable = true)\n",
      " |    |-- element: string (containsNull = true)\n",
      " |-- genomicLocation: struct (nullable = true)\n",
      " |    |-- chromosome: string (nullable = true)\n",
      " |    |-- start: long (nullable = true)\n",
      " |    |-- end: long (nullable = true)\n",
      " |    |-- strand: integer (nullable = true)\n",
      " |-- alternativeGenes: array (nullable = true)\n",
      " |    |-- element: string (containsNull = true)\n",
      " |-- approvedName: string (nullable = true)\n",
      " |-- go: array (nullable = true)\n",
      " |    |-- element: struct (containsNull = true)\n",
      " |    |    |-- id: string (nullable = true)\n",
      " |    |    |-- source: string (nullable = true)\n",
      " |    |    |-- evidence: string (nullable = true)\n",
      " |    |    |-- aspect: string (nullable = true)\n",
      " |    |    |-- geneProduct: string (nullable = true)\n",
      " |    |    |-- ecoId: string (nullable = true)\n",
      " |-- hallmarks: struct (nullable = true)\n",
      " |    |-- attributes: array (nullable = true)\n",
      " |    |    |-- element: struct (containsNull = true)\n",
      " |    |    |    |-- pmid: long (nullable = true)\n",
      " |    |    |    |-- description: string (nullable = true)\n",
      " |    |    |    |-- attribute_name: string (nullable = true)\n",
      " |    |-- cancerHallmarks: array (nullable = true)\n",
      " |    |    |-- element: struct (containsNull = true)\n",
      " |    |    |    |-- pmid: long (nullable = true)\n",
      " |    |    |    |-- description: string (nullable = true)\n",
      " |    |    |    |-- impact: string (nullable = true)\n",
      " |    |    |    |-- label: string (nullable = true)\n",
      " |-- synonyms: array (nullable = true)\n",
      " |    |-- element: struct (containsNull = true)\n",
      " |    |    |-- label: string (nullable = true)\n",
      " |    |    |-- source: string (nullable = true)\n",
      " |-- symbolSynonyms: array (nullable = true)\n",
      " |    |-- element: struct (containsNull = true)\n",
      " |    |    |-- label: string (nullable = true)\n",
      " |    |    |-- source: string (nullable = true)\n",
      " |-- nameSynonyms: array (nullable = true)\n",
      " |    |-- element: struct (containsNull = true)\n",
      " |    |    |-- label: string (nullable = true)\n",
      " |    |    |-- source: string (nullable = true)\n",
      " |-- functionDescriptions: array (nullable = true)\n",
      " |    |-- element: string (containsNull = true)\n",
      " |-- subcellularLocations: array (nullable = true)\n",
      " |    |-- element: struct (containsNull = true)\n",
      " |    |    |-- location: string (nullable = true)\n",
      " |    |    |-- source: string (nullable = true)\n",
      " |    |    |-- termSL: string (nullable = true)\n",
      " |    |    |-- labelSL: string (nullable = true)\n",
      " |-- targetClass: array (nullable = true)\n",
      " |    |-- element: struct (containsNull = true)\n",
      " |    |    |-- id: long (nullable = true)\n",
      " |    |    |-- label: string (nullable = true)\n",
      " |    |    |-- level: string (nullable = true)\n",
      " |-- obsoleteSymbols: array (nullable = true)\n",
      " |    |-- element: struct (containsNull = true)\n",
      " |    |    |-- label: string (nullable = true)\n",
      " |    |    |-- source: string (nullable = true)\n",
      " |-- obsoleteNames: array (nullable = true)\n",
      " |    |-- element: struct (containsNull = true)\n",
      " |    |    |-- label: string (nullable = true)\n",
      " |    |    |-- source: string (nullable = true)\n",
      " |-- constraint: array (nullable = true)\n",
      " |    |-- element: struct (containsNull = true)\n",
      " |    |    |-- constraintType: string (nullable = true)\n",
      " |    |    |-- score: float (nullable = true)\n",
      " |    |    |-- exp: float (nullable = true)\n",
      " |    |    |-- obs: integer (nullable = true)\n",
      " |    |    |-- oe: float (nullable = true)\n",
      " |    |    |-- oeLower: float (nullable = true)\n",
      " |    |    |-- oeUpper: float (nullable = true)\n",
      " |    |    |-- upperRank: integer (nullable = true)\n",
      " |    |    |-- upperBin: integer (nullable = true)\n",
      " |    |    |-- upperBin6: integer (nullable = true)\n",
      " |-- tep: struct (nullable = true)\n",
      " |    |-- targetFromSourceId: string (nullable = true)\n",
      " |    |-- description: string (nullable = true)\n",
      " |    |-- therapeuticArea: string (nullable = true)\n",
      " |    |-- url: string (nullable = true)\n",
      " |-- proteinIds: array (nullable = true)\n",
      " |    |-- element: struct (containsNull = true)\n",
      " |    |    |-- id: string (nullable = true)\n",
      " |    |    |-- source: string (nullable = true)\n",
      " |-- dbXrefs: array (nullable = true)\n",
      " |    |-- element: struct (containsNull = true)\n",
      " |    |    |-- id: string (nullable = true)\n",
      " |    |    |-- source: string (nullable = true)\n",
      " |-- chemicalProbes: array (nullable = true)\n",
      " |    |-- element: struct (containsNull = true)\n",
      " |    |    |-- control: string (nullable = true)\n",
      " |    |    |-- drugId: string (nullable = true)\n",
      " |    |    |-- id: string (nullable = true)\n",
      " |    |    |-- isHighQuality: boolean (nullable = true)\n",
      " |    |    |-- mechanismOfAction: array (nullable = true)\n",
      " |    |    |    |-- element: string (containsNull = true)\n",
      " |    |    |-- origin: array (nullable = true)\n",
      " |    |    |    |-- element: string (containsNull = true)\n",
      " |    |    |-- probeMinerScore: long (nullable = true)\n",
      " |    |    |-- probesDrugsScore: long (nullable = true)\n",
      " |    |    |-- scoreInCells: long (nullable = true)\n",
      " |    |    |-- scoreInOrganisms: long (nullable = true)\n",
      " |    |    |-- targetFromSourceId: string (nullable = true)\n",
      " |    |    |-- urls: array (nullable = true)\n",
      " |    |    |    |-- element: struct (containsNull = true)\n",
      " |    |    |    |    |-- niceName: string (nullable = true)\n",
      " |    |    |    |    |-- url: string (nullable = true)\n",
      " |-- homologues: array (nullable = true)\n",
      " |    |-- element: struct (containsNull = true)\n",
      " |    |    |-- speciesId: string (nullable = true)\n",
      " |    |    |-- speciesName: string (nullable = true)\n",
      " |    |    |-- homologyType: string (nullable = true)\n",
      " |    |    |-- targetGeneId: string (nullable = true)\n",
      " |    |    |-- isHighConfidence: string (nullable = true)\n",
      " |    |    |-- targetGeneSymbol: string (nullable = true)\n",
      " |    |    |-- queryPercentageIdentity: double (nullable = true)\n",
      " |    |    |-- targetPercentageIdentity: double (nullable = true)\n",
      " |    |    |-- priority: integer (nullable = true)\n",
      " |-- tractability: array (nullable = true)\n",
      " |    |-- element: struct (containsNull = true)\n",
      " |    |    |-- modality: string (nullable = true)\n",
      " |    |    |-- id: string (nullable = true)\n",
      " |    |    |-- value: boolean (nullable = true)\n",
      " |-- safetyLiabilities: array (nullable = true)\n",
      " |    |-- element: struct (containsNull = true)\n",
      " |    |    |-- event: string (nullable = true)\n",
      " |    |    |-- eventId: string (nullable = true)\n",
      " |    |    |-- effects: array (nullable = true)\n",
      " |    |    |    |-- element: struct (containsNull = true)\n",
      " |    |    |    |    |-- direction: string (nullable = true)\n",
      " |    |    |    |    |-- dosing: string (nullable = true)\n",
      " |    |    |-- biosamples: array (nullable = true)\n",
      " |    |    |    |-- element: struct (containsNull = true)\n",
      " |    |    |    |    |-- cellFormat: string (nullable = true)\n",
      " |    |    |    |    |-- cellLabel: string (nullable = true)\n",
      " |    |    |    |    |-- tissueId: string (nullable = true)\n",
      " |    |    |    |    |-- tissueLabel: string (nullable = true)\n",
      " |    |    |-- datasource: string (nullable = true)\n",
      " |    |    |-- literature: string (nullable = true)\n",
      " |    |    |-- url: string (nullable = true)\n",
      " |    |    |-- studies: array (nullable = true)\n",
      " |    |    |    |-- element: struct (containsNull = true)\n",
      " |    |    |    |    |-- description: string (nullable = true)\n",
      " |    |    |    |    |-- name: string (nullable = true)\n",
      " |    |    |    |    |-- type: string (nullable = true)\n",
      " |-- pathways: array (nullable = true)\n",
      " |    |-- element: struct (containsNull = true)\n",
      " |    |    |-- pathwayId: string (nullable = true)\n",
      " |    |    |-- pathway: string (nullable = true)\n",
      " |    |    |-- topLevelTerm: string (nullable = true)\n",
      " |-- tss: long (nullable = true)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "target.printSchema()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
